{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e6dd9a48-1292-4cb6-a417-c7d7af76a8cc",
   "metadata": {},
   "source": [
    "### 1. Create SCENIC+ object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c25cdda4-08bc-483a-a61c-3b5b7dd1df32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load functions\n",
    "from scenicplus.scenicplus_class import SCENICPLUS, create_SCENICPLUS_object\n",
    "from scenicplus.preprocessing.filtering import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f130f5ac-c88b-466e-872c-f3b81859b851",
   "metadata": {},
   "source": [
    "First we will load the scRNA-seq and the scATAC-seq data. We make sure that names match between them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f2ec5693-afaa-41aa-b565-2a04a1e3e0b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local_scratch/tmp-vsc31305/ipykernel_10949/1356271359.py:6: DeprecationWarning: Please use `csr_matrix` from the `scipy.sparse` namespace, the `scipy.sparse.csr` namespace is deprecated.\n",
      "  cistopic_obj = pickle.load(infile)\n",
      "/local_scratch/tmp-vsc31305/ipykernel_10949/1356271359.py:11: DeprecationWarning: Please use `csr_matrix` from the `scipy.sparse` namespace, the `scipy.sparse.csr` namespace is deprecated.\n",
      "  imputed_acc_obj = pickle.load(infile)\n"
     ]
    }
   ],
   "source": [
    "# Load data\n",
    "## ATAC - cisTopic object\n",
    "outDir = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/pycisTopic/'\n",
    "import pickle\n",
    "infile = open(outDir + 'DPCL_cisTopicObject.pkl', 'rb')\n",
    "cistopic_obj = pickle.load(infile)\n",
    "infile.close()\n",
    "## Precomputed imputed data\n",
    "import pickle\n",
    "infile = open(outDir + 'DARs/Imputed_accessibility.pkl', 'rb')\n",
    "imputed_acc_obj = pickle.load(infile)\n",
    "infile.close()\n",
    "## RNA - Create Anndata\n",
    "from loomxpy.loomxpy import SCopeLoom\n",
    "from pycisTopic.loom import *\n",
    "import itertools\n",
    "import anndata\n",
    "path_to_loom = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/vsn/grnboost/out/data/scRNA_count_matrix.SINGLE_SAMPLE_SCENIC.loom'\n",
    "loom = SCopeLoom.read_loom(path_to_loom)\n",
    "metadata = get_metadata(loom)\n",
    "# Fix names\n",
    "metadata.index =  metadata.index + '___DPLC'\n",
    "expr_mat = loom.ex_mtx\n",
    "expr_mat.index =  expr_mat.index + '___DPLC'\n",
    "rna_anndata = anndata.AnnData(X=expr_mat)\n",
    "rna_anndata.obs = metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83f0eb5d-e2e3-448f-9a78-9fb3c9d7b3db",
   "metadata": {},
   "source": [
    "Next we load the motif enrichment results into a dictionary. We can load motif results from the different methods in pycistarget (e.g. cisTarget, DEM) and different region sets (e.g. topics, DARs, MACS bdgdiff peaks). In this tutorial we will use both cisTarget and DEM peaks from topics and DARs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "62e60c6f-21c6-4a7c-a720-83c042c12106",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Precomputed imputed data\n",
    "import pickle\n",
    "infile = open('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/pycistarget/cluster_V10_V2/menr.pkl', 'rb') \n",
    "menr = pickle.load(infile)\n",
    "infile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a801f29-be66-4c2c-b061-0c9789d94663",
   "metadata": {},
   "source": [
    "Now we can create the SCENIC+ object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0e1a159-8d69-4fec-891c-a05e081e4980",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj = create_SCENICPLUS_object(\n",
    "        GEX_anndata = rna_anndata,\n",
    "        cisTopic_obj = cistopic_obj,\n",
    "        imputed_acc_obj = imputed_acc_obj,\n",
    "        menr = menr,\n",
    "        ACC_prefix = 'ACC_',\n",
    "        GEX_prefix = 'GEX_',\n",
    "        bc_transform_func = lambda x: x,\n",
    "        normalize_imputed_acc = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4a7ec27-2e3c-4563-9550-f085247a4adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "type(scplus_obj.X_EXP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b99202c-97cb-49f3-a8c6-f59adbe87e9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SCENIC+ object with n_cells x n_genes = 4000 x 33415 and n_cells x n_regions = 4000 x 631497\n",
      "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
      "\tmetadata_cell:'GEX_sample_id', 'GEX_leiden', 'ACC_cisTopic_nr_frag', 'ACC_cisTopic_log_nr_frag', 'ACC_cisTopic_nr_acc', 'ACC_cisTopic_log_nr_acc', 'ACC_sample_id', 'ACC_Log_total_nr_frag', 'ACC_Log_unique_nr_frag', 'ACC_Total_nr_frag', 'ACC_Unique_nr_frag', 'ACC_Dupl_nr_frag', 'ACC_Dupl_rate', 'ACC_Total_nr_frag_in_regions', 'ACC_Unique_nr_frag_in_regions', 'ACC_FRIP', 'ACC_TSS_enrichment', 'ACC_barcode', 'ACC_Cell_type', 'ACC_pycisTopic_leiden_10_0.3', 'ACC_Line_type'\n",
      "\tmenr:'CTX_Topics_All', 'CTX_Topics_NP', 'CTX_DARs_All', 'CTX_DARs_NP', 'DEM_Topics_All', 'DEM_Topics_NP', 'DEM_DARs_All', 'DEM_DARs_NP'\n",
      "\tdr_cell:'ACC_UMAP', 'ACC_tSNE'\n"
     ]
    }
   ],
   "source": [
    "print(scplus_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dd0d447-6aca-44ed-8396-610cf2096753",
   "metadata": {},
   "source": [
    "You can also filter low accessible regions and low expressed genes. This recommended to avoid getting false relationships with these regions and genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ef7856a8-4173-4889-8ec9-273296f73b22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-01-10 18:27:12,010 Preprocessing INFO     Going from 33415 genes to 25651 genes.\n",
      "2022-01-10 18:28:17,478 Preprocessing INFO     Going from 631497 regions to 589734 regions.\n"
     ]
    }
   ],
   "source": [
    "filter_genes(scplus_obj, min_pct = 0.5)\n",
    "filter_regions(scplus_obj, min_pct = 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5681f5dd-2bad-4970-b223-a3e889b3dfa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "outDir = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final/'\n",
    "import pickle\n",
    "with open(outDir+'scplus_obj.pkl', 'wb') as f:\n",
    "  pickle.dump(scplus_obj, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bded410-ecc9-4fdb-8275-877c10af02dd",
   "metadata": {},
   "source": [
    "### GRNBoost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "562042bf-0944-4285-85e5-8789e03e6579",
   "metadata": {},
   "outputs": [],
   "source": [
    "singularity exec -B /lustre1,/staging,/data,/vsc-hard-mounts,/scratch scenicplus.sif ipython3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "409e7d63-2b43-47e4-b180-fba247398951",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the downstream analyses\n",
    "outDir = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final/'\n",
    "import pickle\n",
    "infile = open(outDir+'scplus_obj.pkl', 'rb')\n",
    "scplus_obj = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "import pickle\n",
    "infile = open('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_V2_grnboost/region_ranking.pkl', 'rb')\n",
    "region_ranking = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "import pickle\n",
    "infile = open('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_V2_grnboost/gene_ranking.pkl', 'rb')\n",
    "gene_ranking = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "from scenicplus.wrappers.run_scenicplus import *\n",
    "run_scenicplus(scplus_obj,\n",
    "    variable = ['ACC_Cell_type'],\n",
    "    species = 'hsapiens',\n",
    "    assembly = 'hg38',\n",
    "    tf_file = '/staging/leuven/stg_00002/lcb/cflerin/resources/allTFs_hg38.txt',\n",
    "    save_path = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final_autoreg/grnboost/',\n",
    "    biomart_host = 'http://oct2016.archive.ensembl.org/',\n",
    "    upstream = [1000, 150000],\n",
    "    downstream = [1000, 150000],   \n",
    "    calculate_TF_eGRN_correlation = False,\n",
    "    calculate_DEGs_DARs = True,\n",
    "    export_to_loom_file = True,\n",
    "    export_to_UCSC_file = True,\n",
    "    region_ranking=region_ranking,\n",
    "    gene_ranking=gene_ranking,\n",
    "    tree_structure = ('DPCL', 'SCENIC+', 'grnboost'),\n",
    "    path_bedToBigBed = '/data/leuven/software/biomed/haswell_centos7/2018a/software/Kent_tools/20190730-linux.x86_64/bin/',\n",
    "    n_cpu = 20,\n",
    "    _temp_dir = '/scratch/leuven/313/vsc31305/ray_spill'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b95c3b29-43d7-40cf-bc85-ecd51945ea03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the downstream analyses\n",
    "outDir = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final_autoreg/grnboost/'\n",
    "import dill\n",
    "infile = open(outDir+'scplus_obj.pkl', 'rb')\n",
    "scplus_obj = dill.load(infile)\n",
    "infile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93f1eac8-72d4-4785-92ca-24010901739d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "def format_egrns_time(scplus_obj,\n",
    "                 eregulons_key: str = 'eRegulons',\n",
    "                 TF2G_key: str = 'TF2G_adj',\n",
    "                 key_added: str = 'eRegulon_metadata'):\n",
    "    \"\"\"\n",
    "    A function to format eRegulons to a pandas dataframe\n",
    "    \"\"\"\n",
    "\n",
    "    egrn_list = scplus_obj.uns[eregulons_key]\n",
    "    TF = [egrn_list[x].transcription_factor for x in range(len(egrn_list))]\n",
    "    is_extended = [str(egrn_list[x].is_extended)\n",
    "                   for x in range(len(egrn_list))]\n",
    "    r2g_data = [pd.DataFrame.from_records(egrn_list[x].regions2genes, columns=[\n",
    "                                          'Region', 'Gene', 'R2G_importance', 'R2G_rho', 'R2G_importance_x_rho', 'R2G_importance_x_abs_rho']) for x in range(len(egrn_list))]\n",
    "    egrn_name = [TF[x] + '_extended' if is_extended[x] ==\n",
    "                 'True' else TF[x] for x in range(len(egrn_list))]\n",
    "    egrn_name = [egrn_name[x] + '_+' if 'positive tf2g' in egrn_list[x]\n",
    "                 .context else egrn_name[x] + '_-' for x in range(len(egrn_list))]\n",
    "    egrn_name = [egrn_name[x] + '_+' if 'positive r2g' in egrn_list[x]\n",
    "                 .context else egrn_name[x] + '_-' for x in range(len(egrn_list))]\n",
    "    region_signature_name = [\n",
    "        egrn_name[x] + '_(' + str(len(set(r2g_data[x].Region))) + 'r)' for x in range(len(egrn_list))]\n",
    "    gene_signature_name = [\n",
    "        egrn_name[x] + '_(' + str(len(set(r2g_data[x].Gene))) + 'g)' for x in range(len(egrn_list))]\n",
    "\n",
    "    for x in range(len(egrn_list)):\n",
    "        r2g_data[x].insert(0, \"TF\", TF[x])\n",
    "        r2g_data[x].insert(1, \"is_extended\", is_extended[x])\n",
    "        r2g_data[x].insert(0, \"Gene_signature_name\", gene_signature_name[x])\n",
    "        r2g_data[x].insert(0, \"Region_signature_name\",\n",
    "                           region_signature_name[x])\n",
    "\n",
    "    tf2g_data = scplus_obj.uns[TF2G_key].copy()\n",
    "    tf2g_data.columns = ['TF', 'Gene', 'TF2G_importance', 'TF2G_regulation',\n",
    "                         'TF2G_rho', 'TF2G_importance_x_abs_rho', 'TF2G_importance_x_rho']\n",
    "    egrn_metadata = pd.concat([pd.merge(r2g_data[x], tf2g_data[tf2g_data.TF == r2g_data[x].TF[0]], on=[\n",
    "                              'TF', 'Gene']) for x in range(len(egrn_list)) if tf2g_data[tf2g_data.TF == r2g_data[x].TF[0]].shape[0] != 0 and r2g_data[x].shape[0] != 0])\n",
    "    scplus_obj.uns[key_added] = egrn_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "190ecb19-6de5-4707-91ec-7ca42013da8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "format_egrns_time(scplus_obj,\n",
    "              eregulons_key = 'eRegulons',\n",
    "              TF2G_key = 'TF2G_adj',\n",
    "              key_added = 'eRegulon_metadata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1ba585a-e57f-4628-a66f-7d2ef6c69912",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj.uns['region_to_gene'].to_csv('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_V2_grnboost_2/region_to_gene.tsv', sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3eb0d17-b9bc-40d4-94f5-8adf392d6a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj.uns['region_to_gene'][scplus_obj.uns['region_to_gene']['rho'] >0].to_csv('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_V2_grnboost_2/region_to_gene_pos.tsv', sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "31d057a8-304d-49ab-8fbf-8f56c5b24957",
   "metadata": {},
   "outputs": [],
   "source": [
    "select = list(set(scplus_obj.uns['eRegulon_metadata']['Region']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d75e0f62-1fdd-4b43-8336-79951a0a93b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>importance</th>\n",
       "      <th>rho</th>\n",
       "      <th>importance_x_rho</th>\n",
       "      <th>importance_x_abs_rho</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13903093-13903594</td>\n",
       "      <td>0.007320</td>\n",
       "      <td>0.039632</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>[-150021]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13916776-13917277</td>\n",
       "      <td>0.001618</td>\n",
       "      <td>-0.008657</td>\n",
       "      <td>-0.000014</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>[-136339]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13918892-13919393</td>\n",
       "      <td>0.007048</td>\n",
       "      <td>0.097071</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>[-134223]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13924873-13925374</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>0.073550</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>[-128241]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13929165-13929666</td>\n",
       "      <td>0.002574</td>\n",
       "      <td>-0.079550</td>\n",
       "      <td>-0.000205</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>[-123949]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083810</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27400009-27400510</td>\n",
       "      <td>0.029715</td>\n",
       "      <td>-0.025321</td>\n",
       "      <td>-0.000752</td>\n",
       "      <td>0.000752</td>\n",
       "      <td>[141840]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083811</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27404043-27404544</td>\n",
       "      <td>0.004020</td>\n",
       "      <td>-0.004116</td>\n",
       "      <td>-0.000017</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>[145874]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083812</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27404765-27405266</td>\n",
       "      <td>0.002344</td>\n",
       "      <td>0.177206</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>[146596]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083813</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27405285-27405786</td>\n",
       "      <td>0.019289</td>\n",
       "      <td>0.195575</td>\n",
       "      <td>0.003772</td>\n",
       "      <td>0.003772</td>\n",
       "      <td>[147116]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083814</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27408123-27408624</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.049839</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[149954]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2083815 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        target                   region  importance       rho  \\\n",
       "0        PALM3  chr19:13903093-13903594    0.007320  0.039632   \n",
       "1        PALM3  chr19:13916776-13917277    0.001618 -0.008657   \n",
       "2        PALM3  chr19:13918892-13919393    0.007048  0.097071   \n",
       "3        PALM3  chr19:13924873-13925374    0.000756  0.073550   \n",
       "4        PALM3  chr19:13929165-13929666    0.002574 -0.079550   \n",
       "...        ...                      ...         ...       ...   \n",
       "2083810  STMN4   chr8:27400009-27400510    0.029715 -0.025321   \n",
       "2083811  STMN4   chr8:27404043-27404544    0.004020 -0.004116   \n",
       "2083812  STMN4   chr8:27404765-27405266    0.002344  0.177206   \n",
       "2083813  STMN4   chr8:27405285-27405786    0.019289  0.195575   \n",
       "2083814  STMN4   chr8:27408123-27408624    0.000000 -0.049839   \n",
       "\n",
       "         importance_x_rho  importance_x_abs_rho   Distance  \n",
       "0                0.000290              0.000290  [-150021]  \n",
       "1               -0.000014              0.000014  [-136339]  \n",
       "2                0.000684              0.000684  [-134223]  \n",
       "3                0.000056              0.000056  [-128241]  \n",
       "4               -0.000205              0.000205  [-123949]  \n",
       "...                   ...                   ...        ...  \n",
       "2083810         -0.000752              0.000752   [141840]  \n",
       "2083811         -0.000017              0.000017   [145874]  \n",
       "2083812          0.000415              0.000415   [146596]  \n",
       "2083813          0.003772              0.003772   [147116]  \n",
       "2083814         -0.000000              0.000000   [149954]  \n",
       "\n",
       "[2083815 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.uns['region_to_gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3d976781-0a87-4e4a-9cd6-6ef5f724038c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>importance</th>\n",
       "      <th>rho</th>\n",
       "      <th>importance_x_rho</th>\n",
       "      <th>importance_x_abs_rho</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13937615-13938116</td>\n",
       "      <td>0.005005</td>\n",
       "      <td>-0.070210</td>\n",
       "      <td>-0.000351</td>\n",
       "      <td>0.000351</td>\n",
       "      <td>[-115499]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13938255-13938756</td>\n",
       "      <td>0.005120</td>\n",
       "      <td>-0.055077</td>\n",
       "      <td>-0.000282</td>\n",
       "      <td>0.000282</td>\n",
       "      <td>[-114859]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13958022-13958523</td>\n",
       "      <td>0.008706</td>\n",
       "      <td>0.088218</td>\n",
       "      <td>0.000768</td>\n",
       "      <td>0.000768</td>\n",
       "      <td>[-95093]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13962328-13962829</td>\n",
       "      <td>0.019312</td>\n",
       "      <td>-0.119996</td>\n",
       "      <td>-0.002317</td>\n",
       "      <td>0.002317</td>\n",
       "      <td>[-90787]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13964715-13965216</td>\n",
       "      <td>0.005130</td>\n",
       "      <td>-0.121418</td>\n",
       "      <td>-0.000623</td>\n",
       "      <td>0.000623</td>\n",
       "      <td>[-88399]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083786</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27363649-27364150</td>\n",
       "      <td>0.028793</td>\n",
       "      <td>0.137378</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>0.003956</td>\n",
       "      <td>[105480]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083787</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27365988-27366489</td>\n",
       "      <td>0.033363</td>\n",
       "      <td>0.189441</td>\n",
       "      <td>0.006320</td>\n",
       "      <td>0.006320</td>\n",
       "      <td>[107818]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083788</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27366506-27367007</td>\n",
       "      <td>0.053182</td>\n",
       "      <td>0.124266</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>[108336]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083796</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27378607-27379108</td>\n",
       "      <td>0.044126</td>\n",
       "      <td>0.199834</td>\n",
       "      <td>0.008818</td>\n",
       "      <td>0.008818</td>\n",
       "      <td>[120438]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083813</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27405285-27405786</td>\n",
       "      <td>0.019289</td>\n",
       "      <td>0.195575</td>\n",
       "      <td>0.003772</td>\n",
       "      <td>0.003772</td>\n",
       "      <td>[147116]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>402836 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        target                   region  importance       rho  \\\n",
       "12       PALM3  chr19:13937615-13938116    0.005005 -0.070210   \n",
       "13       PALM3  chr19:13938255-13938756    0.005120 -0.055077   \n",
       "19       PALM3  chr19:13958022-13958523    0.008706  0.088218   \n",
       "24       PALM3  chr19:13962328-13962829    0.019312 -0.119996   \n",
       "26       PALM3  chr19:13964715-13965216    0.005130 -0.121418   \n",
       "...        ...                      ...         ...       ...   \n",
       "2083786  STMN4   chr8:27363649-27364150    0.028793  0.137378   \n",
       "2083787  STMN4   chr8:27365988-27366489    0.033363  0.189441   \n",
       "2083788  STMN4   chr8:27366506-27367007    0.053182  0.124266   \n",
       "2083796  STMN4   chr8:27378607-27379108    0.044126  0.199834   \n",
       "2083813  STMN4   chr8:27405285-27405786    0.019289  0.195575   \n",
       "\n",
       "         importance_x_rho  importance_x_abs_rho   Distance  \n",
       "12              -0.000351              0.000351  [-115499]  \n",
       "13              -0.000282              0.000282  [-114859]  \n",
       "19               0.000768              0.000768   [-95093]  \n",
       "24              -0.002317              0.002317   [-90787]  \n",
       "26              -0.000623              0.000623   [-88399]  \n",
       "...                   ...                   ...        ...  \n",
       "2083786          0.003956              0.003956   [105480]  \n",
       "2083787          0.006320              0.006320   [107818]  \n",
       "2083788          0.006609              0.006609   [108336]  \n",
       "2083796          0.008818              0.008818   [120438]  \n",
       "2083813          0.003772              0.003772   [147116]  \n",
       "\n",
       "[402836 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.uns['region_to_gene'][scplus_obj.uns['region_to_gene']['region'].isin(select)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "06fd4618-68ef-44f8-b5b1-009e8a82660c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'scplus_obj' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [1]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mscplus_obj\u001b[49m\u001b[38;5;241m.\u001b[39muns[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mregion_to_gene\u001b[39m\u001b[38;5;124m'\u001b[39m][scplus_obj\u001b[38;5;241m.\u001b[39muns[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mregion_to_gene\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mregion\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(select)]\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_V2_grnboost_2/region_to_gene_in_eGRN.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'scplus_obj' is not defined"
     ]
    }
   ],
   "source": [
    "scplus_obj.uns['region_to_gene'][scplus_obj.uns['region_to_gene']['region'].isin(select)].to_csv('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_V2_grnboost_2/region_to_gene_in_eGRN.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e78812d1-132f-450d-857b-287b373a6db6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>importance</th>\n",
       "      <th>rho</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21587750-21588251</td>\n",
       "      <td>10.000017</td>\n",
       "      <td>10.000017</td>\n",
       "      <td>190631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21585171-21585672</td>\n",
       "      <td>10.000017</td>\n",
       "      <td>10.000017</td>\n",
       "      <td>188052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21586234-21586735</td>\n",
       "      <td>10.000017</td>\n",
       "      <td>10.000017</td>\n",
       "      <td>189115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21282787-21283288</td>\n",
       "      <td>10.034795</td>\n",
       "      <td>10.034795</td>\n",
       "      <td>-114332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21221605-21222106</td>\n",
       "      <td>10.049212</td>\n",
       "      <td>10.049212</td>\n",
       "      <td>-175514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576842</th>\n",
       "      <td>NPIPA2</td>\n",
       "      <td>chr16:14680470-14680971</td>\n",
       "      <td>3856.234900</td>\n",
       "      <td>3856.234900</td>\n",
       "      <td>-67596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576843</th>\n",
       "      <td>SLX1A</td>\n",
       "      <td>chr16:22174595-22175096</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>-8019280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576844</th>\n",
       "      <td>SLX1A</td>\n",
       "      <td>chr16:22170680-22171181</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>-8023195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576845</th>\n",
       "      <td>BOLA2B</td>\n",
       "      <td>chr16:22174595-22175096</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>8019711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576846</th>\n",
       "      <td>BOLA2B</td>\n",
       "      <td>chr16:22170680-22171181</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>8023626</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2319660 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         target                   region   importance          rho  Distance\n",
       "0        ZNF493  chr19:21587750-21588251    10.000017    10.000017    190631\n",
       "1        ZNF493  chr19:21585171-21585672    10.000017    10.000017    188052\n",
       "2        ZNF493  chr19:21586234-21586735    10.000017    10.000017    189115\n",
       "3        ZNF493  chr19:21282787-21283288    10.034795    10.034795   -114332\n",
       "4        ZNF493  chr19:21221605-21222106    10.049212    10.049212   -175514\n",
       "...         ...                      ...          ...          ...       ...\n",
       "2576842  NPIPA2  chr16:14680470-14680971  3856.234900  3856.234900    -67596\n",
       "2576843   SLX1A  chr16:22174595-22175096  3885.789000  3885.789000  -8019280\n",
       "2576844   SLX1A  chr16:22170680-22171181  3885.789000  3885.789000  -8023195\n",
       "2576845  BOLA2B  chr16:22174595-22175096  3885.789000  3885.789000   8019711\n",
       "2576846  BOLA2B  chr16:22170680-22171181  3885.789000  3885.789000   8023626\n",
       "\n",
       "[2319660 rows x 5 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "hic_data = pd.read_csv('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/formatted/HepG2_ENCFF020DPP_5Kb_SCALE.txt', sep='\\t')\n",
    "hic_data.columns = ['Chromosome', 'Start', 'End', 'target', 'importance', 'TSS', 'Strand', 'Distance', 'region']\n",
    "hic_data['rho'] = hic_data['importance']\n",
    "hic_data = hic_data[['target', 'region', 'importance', 'rho', 'Distance']]\n",
    "hic_data = hic_data[hic_data['region'].isin(scplus_obj.region_names)]\n",
    "hic_data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c97fdd23-bce5-4407-95e8-0b0374e166a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import ray\n",
    "import logging\n",
    "import time\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "import pyranges as pr\n",
    "from sklearn.ensemble import GradientBoostingRegressor, RandomForestRegressor, ExtraTreesRegressor\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "from tqdm import tqdm\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import Normalize\n",
    "from typing import List\n",
    "\n",
    "from scenicplus.utils import extend_pyranges, extend_pyranges_with_limits, reduce_pyranges_with_limits_b\n",
    "from scenicplus.utils import calculate_distance_with_limits_join, reduce_pyranges_b, calculate_distance_join\n",
    "from scenicplus.utils import coord_to_region_names, region_names_to_coordinates, ASM_SYNONYMS, Groupby, flatten_list\n",
    "from scenicplus.scenicplus_class import SCENICPLUS\n",
    "from scenicplus.enhancer_to_gene import INTERACT_AS\n",
    "\n",
    "\n",
    "\n",
    "def export_to_UCSC_interact_hic(SCENICPLUS_obj: SCENICPLUS,\n",
    "                            species: str,\n",
    "                            outfile: str,\n",
    "                            region_to_gene_key: str =' region_to_gene',\n",
    "                            pbm_host:str = 'http://www.ensembl.org',\n",
    "                            bigbed_outfile:str = None,\n",
    "                            path_bedToBigBed: str= None,\n",
    "                            assembly: str = None,\n",
    "                            ucsc_track_name: str = 'region_to_gene',\n",
    "                            ucsc_description: str = 'interaction file for region to gene',\n",
    "                            cmap_neg: str = 'Reds',\n",
    "                            cmap_pos: str = 'Greens',\n",
    "                            key_for_color: str = 'importance',\n",
    "                            vmin: int = 0,\n",
    "                            vmax: int = 1,\n",
    "                            scale_by_gene: bool = True,\n",
    "                            subset_for_eRegulons_regions: bool = True,\n",
    "                            eRegulons_key: str = 'eRegulons') -> pd.DataFrame:\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    format = '%(asctime)s %(name)-12s %(levelname)-8s %(message)s'\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=format, handlers=handlers)\n",
    "    log = logging.getLogger('R2G')\n",
    "\n",
    "    if region_to_gene_key not in SCENICPLUS_obj.uns.keys():\n",
    "        raise Exception(\n",
    "            f'key {region_to_gene_key} not found in SCENICPLUS_obj.uns, first calculate region to gene relationships using function: \"calculate_regions_to_genes_relationships\"')\n",
    "\n",
    "    region_to_gene_df = SCENICPLUS_obj.uns[region_to_gene_key].copy()\n",
    "\n",
    "    if subset_for_eRegulons_regions:\n",
    "        if eRegulons_key not in SCENICPLUS_obj.uns.keys():\n",
    "            raise ValueError(\n",
    "                f'key {eRegulons_key} not found in SCENICPLUS_obj.uns.keys()')\n",
    "        eRegulon_regions = list(set(flatten_list(\n",
    "            [ereg.target_regions for ereg in SCENICPLUS_obj.uns[eRegulons_key]])))\n",
    "        region_to_gene_df.index = region_to_gene_df['region']\n",
    "        region_to_gene_df = region_to_gene_df.loc[eRegulon_regions].reset_index(\n",
    "            drop=True)\n",
    "\n",
    "    # Rename columns to be in line with biomart annotation\n",
    "    region_to_gene_df.rename(columns={'target': 'Gene'}, inplace=True)\n",
    "\n",
    "    # Get TSS annotation (end-point for links)\n",
    "    log.info('Downloading gene annotation from biomart, using dataset: {}'.format(\n",
    "        species+'_gene_ensembl'))\n",
    "    import pybiomart as pbm\n",
    "    dataset = pbm.Dataset(name=species+'_gene_ensembl',  host=pbm_host)\n",
    "    annot = dataset.query(attributes=['chromosome_name', 'start_position', 'end_position',\n",
    "                          'strand', 'external_gene_name', 'transcription_start_site', 'transcript_biotype'])\n",
    "    annot.columns = ['Chromosome', 'Start', 'End', 'Strand',\n",
    "                     'Gene', 'Transcription_Start_Site', 'Transcript_type']\n",
    "    annot['Chromosome'] = 'chr' + \\\n",
    "        annot['Chromosome'].astype(str)\n",
    "    annot = annot[annot.Transcript_type == 'protein_coding']\n",
    "    annot.Strand[annot.Strand == 1] = '+'\n",
    "    annot.Strand[annot.Strand == -1] = '-'\n",
    "\n",
    "    log.info('Formatting data ...')\n",
    "    # get gene to tss mapping, take the one equal to the gene start/end location if possible otherwise take the first one\n",
    "    annot['TSSeqStartEnd'] = np.logical_or(\n",
    "        annot['Transcription_Start_Site'] == annot['Start'], annot['Transcription_Start_Site'] == annot['End'])\n",
    "    gene_to_tss = annot[['Gene', 'Transcription_Start_Site']].groupby(\n",
    "        'Gene').agg(lambda x: list(map(str, x)))\n",
    "    startEndEq = annot[['Gene', 'TSSeqStartEnd']\n",
    "                       ].groupby('Gene').agg(lambda x: list(x))\n",
    "    gene_to_tss['Transcription_Start_Site'] = [np.array(tss[0])[eq[0]][0] if sum(\n",
    "        eq[0]) >= 1 else tss[0][0] for eq, tss in zip(startEndEq.values, gene_to_tss.values)]\n",
    "    gene_to_tss.columns = ['TSS_Gene']\n",
    "\n",
    "    # get gene to strand mapping\n",
    "    gene_to_strand = annot[['Gene', 'Strand']].groupby(\n",
    "        'Gene').agg(lambda x: list(map(str, x))[0])\n",
    "\n",
    "    # get gene to chromosome mapping (should be the same as the regions mapped to the gene)\n",
    "    gene_to_chrom = annot[['Gene', 'Chromosome']].groupby(\n",
    "        'Gene').agg(lambda x: list(map(str, x))[0])\n",
    "\n",
    "    # add TSS for each gene to region_to_gene_df\n",
    "    region_to_gene_df = region_to_gene_df.join(gene_to_tss, on='Gene')\n",
    "\n",
    "    # add strand for each gene to region_to_gene_df\n",
    "    region_to_gene_df = region_to_gene_df.join(gene_to_strand, on='Gene')\n",
    "\n",
    "    # add chromosome for each gene to region_to_gene_df\n",
    "    region_to_gene_df = region_to_gene_df.join(gene_to_chrom, on='Gene')\n",
    "\n",
    "    # get chrom, chromStart, chromEnd\n",
    "    region_to_gene_df.dropna(axis=0, how='any', inplace=True)\n",
    "    arr = region_names_to_coordinates(region_to_gene_df['region']).to_numpy()\n",
    "    chrom, chromStart, chromEnd = np.split(arr, 3, 1)\n",
    "    chrom = chrom[:, 0]\n",
    "    chromStart = chromStart[:, 0]\n",
    "    chromEnd = chromEnd[:, 0]\n",
    "\n",
    "    # get source chrom, chromStart, chromEnd (i.e. middle of regions)\n",
    "    sourceChrom = chrom\n",
    "    sourceStart = np.array(\n",
    "        list(map(int, chromStart + (chromEnd - chromStart)/2 - 1)))\n",
    "    sourceEnd = np.array(\n",
    "        list(map(int, chromStart + (chromEnd - chromStart)/2)))\n",
    "\n",
    "    # get target chrom, chromStart, chromEnd (i.e. TSS)\n",
    "    targetChrom = region_to_gene_df['Chromosome']\n",
    "    targetStart = region_to_gene_df['TSS_Gene'].values\n",
    "    targetEnd = list(map(str, np.array(list(map(int, targetStart))) + np.array(\n",
    "        [1 if strand == '+' else -1 for strand in region_to_gene_df['Strand'].values])))\n",
    "\n",
    "    # get color\n",
    "    norm = Normalize(vmin=vmin, vmax=vmax)\n",
    "    if scale_by_gene:\n",
    "        grouper = Groupby(\n",
    "            region_to_gene_df.loc[:, 'Gene'].to_numpy())\n",
    "        scores = region_to_gene_df.loc[:, key_for_color].to_numpy()\n",
    "        mapper = cm.ScalarMappable(norm=norm, cmap=cmap_pos)\n",
    "\n",
    "        def _value_to_color(scores):\n",
    "            S = (scores - scores.min()) / (scores.max() - scores.min())\n",
    "            return [','.join([str(x) for x in mapper.to_rgba(s, bytes=True)][0:3]) for s in S]\n",
    "\n",
    "        colors_pos = np.zeros(len(scores), dtype='object')\n",
    "        for idx in grouper.indices:\n",
    "            colors_pos[idx] = _value_to_color(scores[idx])\n",
    "\n",
    "        def _value_to_color(scores):\n",
    "            S = (scores - scores.min()) / (scores.max() - scores.min())\n",
    "            return [','.join([str(x) for x in mapper.to_rgba(s, bytes=True)][0:3]) for s in S]\n",
    "\n",
    "        colors_neg = np.zeros(len(scores), dtype='object')\n",
    "        for idx in grouper.indices:\n",
    "            colors_neg[idx] = _value_to_color(scores[idx])\n",
    "\n",
    "    else:\n",
    "        scores = region_to_gene_df.loc[:, key_for_color].to_numpy()\n",
    "        mapper = cm.ScalarMappable(norm=norm, cmap=cmap_pos)\n",
    "        colors_pos = [\n",
    "            ','.join([str(x) for x in mapper.to_rgba(s, bytes=True)][0:3]) for s in scores]\n",
    "\n",
    "\n",
    "    region_to_gene_df.loc[:, 'color'] = colors_pos\n",
    "    region_to_gene_df['color'] = region_to_gene_df['color'].fillna('55,55,55')\n",
    "\n",
    "    # get name for regions (add incremental number to gene in range of regions linked to gene)\n",
    "    counter = 1\n",
    "    previous_gene = region_to_gene_df['Gene'].values[0]\n",
    "    names = []\n",
    "    for gene in region_to_gene_df['Gene'].values:\n",
    "        if gene != previous_gene:\n",
    "            counter = 1\n",
    "        else:\n",
    "            counter += 1\n",
    "        names.append(gene + '_' + str(counter))\n",
    "        previous_gene = gene\n",
    "\n",
    "    # format final interact dataframe\n",
    "    df_interact = pd.DataFrame(\n",
    "        data={\n",
    "            'chrom':        chrom,\n",
    "            'chromStart':   chromStart,\n",
    "            'chromEnd':     chromEnd,\n",
    "            'name':         names,\n",
    "            'score':        (1000*(region_to_gene_df['importance'].values - np.min(region_to_gene_df['importance'].values))/np.ptp(region_to_gene_df['importance'].values)).astype(int) ,\n",
    "            'value':        region_to_gene_df['importance'].values,\n",
    "            'exp':          np.repeat('.', len(region_to_gene_df)),\n",
    "            'color':        region_to_gene_df['color'].values,\n",
    "            'sourceChrom':  sourceChrom,\n",
    "            'sourceStart':  sourceStart,\n",
    "            'sourceEnd':    sourceEnd,\n",
    "            'sourceName':   names,\n",
    "            'sourceStrand': np.repeat('.', len(region_to_gene_df)),\n",
    "            'targetChrom':  targetChrom,\n",
    "            'targetStart':  targetStart,\n",
    "            'targetEnd':    targetEnd,\n",
    "            'targetName':   region_to_gene_df['Gene'].values,\n",
    "            'targetStrand': region_to_gene_df['Strand'].values\n",
    "        }\n",
    "    )\n",
    "    # sort dataframe\n",
    "    df_interact = df_interact.sort_values(by=['chrom', 'chromStart'])\n",
    "    # Write interact file\n",
    "    log.info('Writing data to: {}'.format(outfile))\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write('track type=interact name=\"{}\" description=\"{}\" useScore=0 maxHeightPixels=200:100:50 visibility=full\\n'.format(\n",
    "            ucsc_track_name, ucsc_description))\n",
    "        df_interact.to_csv(f, header=False, index=False, sep='\\t')\n",
    "\n",
    "    # write bigInteract file\n",
    "    if bigbed_outfile != None:\n",
    "        log.info('Writing data to: {}'.format(bigbed_outfile))\n",
    "        outfolder = bigbed_outfile.rsplit('/', 1)[0]\n",
    "        # write bed file without header to tmp file\n",
    "        df_interact.to_csv(os.path.join(\n",
    "            outfolder, 'interact.bed.tmp'), header=False, index=False, sep='\\t')\n",
    "\n",
    "        # check if auto sql definition for interaction file exists in outfolder, otherwise create it\n",
    "        if not os.path.exists(os.path.join(outfolder, 'interact.as')):\n",
    "            with open(os.path.join(outfolder, 'interact.as'), 'w') as f:\n",
    "                f.write(INTERACT_AS)\n",
    "        # convert interact.bed.tmp to bigBed format\n",
    "        # bedToBigBed -as=interact.as -type=bed5+13 region_to_gene_no_head.interact https://genome.ucsc.edu/goldenPath/help/hg38.chrom.sizes region_to_gene.inter.bb\n",
    "        cmds = [\n",
    "            os.path.join(path_bedToBigBed, 'bedToBigBed'),\n",
    "            '-as={}'.format(os.path.join(os.path.join(outfolder, 'interact.as'))),\n",
    "            '-type=bed5+13',\n",
    "            os.path.join(outfolder, 'interact.bed.tmp'),\n",
    "            'https://hgdownload.cse.ucsc.edu/goldenpath/' + assembly + '/bigZips/' + assembly + '.chrom.sizes',\n",
    "            bigbed_outfile\n",
    "        ]\n",
    "        p = subprocess.Popen(cmds, stdout=subprocess.PIPE,\n",
    "                             stderr=subprocess.PIPE)\n",
    "        stdout, stderr = p.communicate()\n",
    "        if p.returncode:\n",
    "            raise ValueError(\n",
    "                \"cmds: %s\\nstderr:%s\\nstdout:%s\" % (\n",
    "                    \" \".join(cmds), stderr, stdout)\n",
    "            )\n",
    "    return df_interact"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c441075e-7cb4-4c23-be5d-7762a1da27f2",
   "metadata": {},
   "source": [
    "# Generate Hi-C files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b1f62fa5-dfaf-4c12-b167-7c4c33aaffd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the downstream analyses\n",
    "outDir = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final_autoreg/grnboost/'\n",
    "import dill\n",
    "infile = open(outDir+'scplus_obj.pkl', 'rb')\n",
    "scplus_obj = dill.load(infile)\n",
    "infile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e3300bb8-0943-4a91-9693-63b171cc7345",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>importance</th>\n",
       "      <th>rho</th>\n",
       "      <th>importance_x_rho</th>\n",
       "      <th>importance_x_abs_rho</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13903093-13903594</td>\n",
       "      <td>0.007320</td>\n",
       "      <td>0.039632</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>[-150021]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13918892-13919393</td>\n",
       "      <td>0.007048</td>\n",
       "      <td>0.097071</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>[-134223]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13924873-13925374</td>\n",
       "      <td>0.000756</td>\n",
       "      <td>0.073550</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>[-128241]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13934093-13934594</td>\n",
       "      <td>0.003016</td>\n",
       "      <td>0.046918</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>[-119021]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PALM3</td>\n",
       "      <td>chr19:13940538-13941039</td>\n",
       "      <td>0.010659</td>\n",
       "      <td>0.113388</td>\n",
       "      <td>0.001209</td>\n",
       "      <td>0.001209</td>\n",
       "      <td>[-112577]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083802</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27388148-27388649</td>\n",
       "      <td>0.009306</td>\n",
       "      <td>0.204597</td>\n",
       "      <td>0.001904</td>\n",
       "      <td>0.001904</td>\n",
       "      <td>[129978]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083806</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27393594-27394095</td>\n",
       "      <td>0.028735</td>\n",
       "      <td>0.188804</td>\n",
       "      <td>0.005425</td>\n",
       "      <td>0.005425</td>\n",
       "      <td>[135424]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083807</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27394954-27395455</td>\n",
       "      <td>0.016020</td>\n",
       "      <td>0.032904</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>[136784]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083812</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27404765-27405266</td>\n",
       "      <td>0.002344</td>\n",
       "      <td>0.177206</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>[146596]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083813</th>\n",
       "      <td>STMN4</td>\n",
       "      <td>chr8:27405285-27405786</td>\n",
       "      <td>0.019289</td>\n",
       "      <td>0.195575</td>\n",
       "      <td>0.003772</td>\n",
       "      <td>0.003772</td>\n",
       "      <td>[147116]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1060220 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        target                   region  importance       rho  \\\n",
       "0        PALM3  chr19:13903093-13903594    0.007320  0.039632   \n",
       "2        PALM3  chr19:13918892-13919393    0.007048  0.097071   \n",
       "3        PALM3  chr19:13924873-13925374    0.000756  0.073550   \n",
       "9        PALM3  chr19:13934093-13934594    0.003016  0.046918   \n",
       "14       PALM3  chr19:13940538-13941039    0.010659  0.113388   \n",
       "...        ...                      ...         ...       ...   \n",
       "2083802  STMN4   chr8:27388148-27388649    0.009306  0.204597   \n",
       "2083806  STMN4   chr8:27393594-27394095    0.028735  0.188804   \n",
       "2083807  STMN4   chr8:27394954-27395455    0.016020  0.032904   \n",
       "2083812  STMN4   chr8:27404765-27405266    0.002344  0.177206   \n",
       "2083813  STMN4   chr8:27405285-27405786    0.019289  0.195575   \n",
       "\n",
       "         importance_x_rho  importance_x_abs_rho   Distance  \n",
       "0                0.000290              0.000290  [-150021]  \n",
       "2                0.000684              0.000684  [-134223]  \n",
       "3                0.000056              0.000056  [-128241]  \n",
       "9                0.000142              0.000142  [-119021]  \n",
       "14               0.001209              0.001209  [-112577]  \n",
       "...                   ...                   ...        ...  \n",
       "2083802          0.001904              0.001904   [129978]  \n",
       "2083806          0.005425              0.005425   [135424]  \n",
       "2083807          0.000527              0.000527   [136784]  \n",
       "2083812          0.000415              0.000415   [146596]  \n",
       "2083813          0.003772              0.003772   [147116]  \n",
       "\n",
       "[1060220 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r2g = scplus_obj.uns['region_to_gene'].copy()\n",
    "r2g = r2g[r2g.rho > 0.03]\n",
    "r2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "51f6a757-bcad-4334-bd30-9b670438168c",
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj.uns['region_to_gene'].to_csv('/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final_autoreg/grnboost/region_to_gene_pos.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08db0345-6eac-42b2-b3ad-5ec370892bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import pandas as pd\n",
    "cell_lines=['GM12878', 'HepG2', 'IMR90', 'HCT116', 'K562']\n",
    "save_path = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/'\n",
    "species = 'hsapiens'\n",
    "assembly = 'hg38'\n",
    "path_bedToBigBed = '/data/leuven/software/biomed/haswell_centos7/2018a/software/Kent_tools/20190730-linux.x86_64/bin/'\n",
    "biomart_host = 'http://oct2016.archive.ensembl.org/'\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/formatted/'\n",
    "for cell_line in cell_lines:\n",
    "    files = [f for f in listdir(path) if isfile(join(path, f))]\n",
    "    file = [f for f in files if cell_line in f]\n",
    "    hic_data = pd.read_csv(path+file[0], sep='\\t')\n",
    "    hic_data.columns = ['Chromosome', 'Start', 'End', 'target', 'importance', 'TSS', 'Strand', 'Distance', 'region']\n",
    "    hic_data['rho'] = hic_data['importance']\n",
    "    hic_data = hic_data[['target', 'region', 'importance', 'rho', 'Distance']]\n",
    "    hic_data = hic_data[hic_data['region'].isin(scplus_obj.region_names)]\n",
    "    scplus_obj.uns['region_to_gene'] = hic_data\n",
    "    r2g_data = export_to_UCSC_interact_hic(scplus_obj,\n",
    "                    species,\n",
    "                    os.path.join(save_path,cell_line+'.hic.all.bed'),\n",
    "                    path_bedToBigBed=path_bedToBigBed,\n",
    "                    bigbed_outfile=os.path.join(save_path,cell_line+'.hic.all.bb'),\n",
    "                    region_to_gene_key='region_to_gene',\n",
    "                    pbm_host=biomart_host,\n",
    "                    assembly=assembly,\n",
    "                    ucsc_track_name='R2G',\n",
    "                    ucsc_description=cell_line+' HiC links',\n",
    "                    cmap_neg='Reds',\n",
    "                    cmap_pos='Greys',\n",
    "                    key_for_color='importance',\n",
    "                    scale_by_gene=True,\n",
    "                    subset_for_eRegulons_regions=False,\n",
    "                    eRegulons_key='eRegulons')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2b6bdf83-acca-4f40-80a8-85276fcec895",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-05-23 13:32:34,106 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-05-23 13:32:34,679 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_29552/570429411.py:143: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n",
      "/tmp/ipykernel_29552/570429411.py:151: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-05-23 13:34:40,211 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/CellOracle.links.all.bed\n",
      "2022-05-23 13:34:42,981 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/CellOracle.links.all.bb\n",
      "2022-05-23 13:34:51,182 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-05-23 13:34:51,617 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_29552/570429411.py:143: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n",
      "/tmp/ipykernel_29552/570429411.py:151: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-05-23 13:35:13,872 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/FigR.links.all.bed\n",
      "2022-05-23 13:35:14,309 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/FigR.links.all.bb\n",
      "2022-05-23 13:35:17,585 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-05-23 13:35:18,122 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_29552/570429411.py:143: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n",
      "/tmp/ipykernel_29552/570429411.py:151: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-05-23 13:36:54,840 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/GRaNIE.links.all.bed\n",
      "2022-05-23 13:36:56,976 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/GRaNIE.links.all.bb\n",
      "2022-05-23 13:37:03,508 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-05-23 13:37:04,247 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_29552/570429411.py:143: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n",
      "/tmp/ipykernel_29552/570429411.py:151: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-05-23 13:38:21,107 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/Scenicplus-importance.links.all.bed\n",
      "2022-05-23 13:38:22,743 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/Scenicplus-importance.links.all.bb\n",
      "2022-05-23 13:38:28,222 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-05-23 13:38:28,760 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_29552/570429411.py:143: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n",
      "/tmp/ipykernel_29552/570429411.py:151: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-05-23 13:39:48,700 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/Scenicplus-rho.links.all.bed\n",
      "2022-05-23 13:39:50,382 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/Scenicplus-rho.links.all.bb\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import pandas as pd\n",
    "cell_lines=['CellOracle', 'FigR', 'GRaNIE', 'Scenicplus-importance', 'Scenicplus-rho']\n",
    "color_dict={'CellOracle':'Greens', 'FigR':'Purples', 'GRaNIE':'Oranges', 'Scenicplus-importance':'Blues', 'Scenicplus-rho':'Reds'}\n",
    "save_path = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/'\n",
    "species = 'hsapiens'\n",
    "assembly = 'hg38'\n",
    "path_bedToBigBed = '/data/leuven/software/biomed/haswell_centos7/2018a/software/Kent_tools/20190730-linux.x86_64/bin/'\n",
    "biomart_host = 'http://oct2016.archive.ensembl.org/'\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/formatted/'\n",
    "for cell_line in cell_lines:\n",
    "    files = [f for f in listdir(path) if isfile(join(path, f))]\n",
    "    file = [f for f in files if cell_line in f]\n",
    "    hic_data = pd.read_csv(path+file[0], sep='\\t')\n",
    "    hic_data.columns = ['Chromosome', 'Start', 'End', 'target', 'importance', 'TSS', 'Strand', 'Distance', 'region']\n",
    "    hic_data['rho'] = hic_data['importance']\n",
    "    hic_data = hic_data[['target', 'region', 'importance', 'rho', 'Distance']]\n",
    "    hic_data = hic_data[hic_data['region'].isin(scplus_obj.region_names)]\n",
    "    scplus_obj.uns['region_to_gene'] = hic_data\n",
    "    r2g_data = export_to_UCSC_interact_hic(scplus_obj,\n",
    "                    species,\n",
    "                    os.path.join(save_path,cell_line+'.links.all.bed'),\n",
    "                    path_bedToBigBed=path_bedToBigBed,\n",
    "                    bigbed_outfile=os.path.join(save_path,cell_line+'.links.all.bb'),\n",
    "                    region_to_gene_key='region_to_gene',\n",
    "                    pbm_host=biomart_host,\n",
    "                    assembly=assembly,\n",
    "                    ucsc_track_name='R2G',\n",
    "                    ucsc_description='SCENIC+ region to gene links',\n",
    "                    cmap_neg='Reds',\n",
    "                    cmap_pos=color_dict[cell_line],\n",
    "                    key_for_color='importance',\n",
    "                    scale_by_gene=True,\n",
    "                    subset_for_eRegulons_regions=False,\n",
    "                    eRegulons_key='eRegulons')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5d5de407-4fd4-448c-b69a-64652771ca87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-05-23 16:33:13,303 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-05-23 16:33:15,676 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_29552/570429411.py:143: RuntimeWarning: invalid value encountered in true_divide\n",
      "  S = (scores - scores.min()) / (scores.max() - scores.min())\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [23]\u001b[0m, in \u001b[0;36m<cell line: 13>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     20\u001b[0m hic_data \u001b[38;5;241m=\u001b[39m hic_data[hic_data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mregion\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misin(scplus_obj\u001b[38;5;241m.\u001b[39mregion_names)]\n\u001b[1;32m     21\u001b[0m scplus_obj\u001b[38;5;241m.\u001b[39muns[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mregion_to_gene\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m hic_data\n\u001b[0;32m---> 22\u001b[0m r2g_data \u001b[38;5;241m=\u001b[39m \u001b[43mexport_to_UCSC_interact_hic\u001b[49m\u001b[43m(\u001b[49m\u001b[43mscplus_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[43m                \u001b[49m\u001b[43mspecies\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     24\u001b[0m \u001b[43m                \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43msave_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcell_line\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.links.all.bed\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     25\u001b[0m \u001b[43m                \u001b[49m\u001b[43mpath_bedToBigBed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpath_bedToBigBed\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     26\u001b[0m \u001b[43m                \u001b[49m\u001b[43mbigbed_outfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43msave_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcell_line\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.links.all.bb\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     27\u001b[0m \u001b[43m                \u001b[49m\u001b[43mregion_to_gene_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mregion_to_gene\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     28\u001b[0m \u001b[43m                \u001b[49m\u001b[43mpbm_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbiomart_host\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[43m                \u001b[49m\u001b[43massembly\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43massembly\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     30\u001b[0m \u001b[43m                \u001b[49m\u001b[43mucsc_track_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mR2G\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[43m                \u001b[49m\u001b[43mucsc_description\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mSCENIC+ region to gene links\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[43m                \u001b[49m\u001b[43mcmap_neg\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mReds\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[43m                \u001b[49m\u001b[43mcmap_pos\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolor_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[43mcell_line\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[43m                \u001b[49m\u001b[43mkey_for_color\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mimportance\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     35\u001b[0m \u001b[43m                \u001b[49m\u001b[43mscale_by_gene\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[43m                \u001b[49m\u001b[43msubset_for_eRegulons_regions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     37\u001b[0m \u001b[43m                \u001b[49m\u001b[43meRegulons_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43meRegulons\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [13]\u001b[0m, in \u001b[0;36mexport_to_UCSC_interact_hic\u001b[0;34m(SCENICPLUS_obj, species, outfile, region_to_gene_key, pbm_host, bigbed_outfile, path_bedToBigBed, assembly, ucsc_track_name, ucsc_description, cmap_neg, cmap_pos, key_for_color, vmin, vmax, scale_by_gene, subset_for_eRegulons_regions, eRegulons_key)\u001b[0m\n\u001b[1;32m    146\u001b[0m colors_pos \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros(\u001b[38;5;28mlen\u001b[39m(scores), dtype\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mobject\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    147\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m idx \u001b[38;5;129;01min\u001b[39;00m grouper\u001b[38;5;241m.\u001b[39mindices:\n\u001b[0;32m--> 148\u001b[0m     colors_pos[idx] \u001b[38;5;241m=\u001b[39m \u001b[43m_value_to_color\u001b[49m\u001b[43m(\u001b[49m\u001b[43mscores\u001b[49m\u001b[43m[\u001b[49m\u001b[43midx\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    150\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_value_to_color\u001b[39m(scores):\n\u001b[1;32m    151\u001b[0m     S \u001b[38;5;241m=\u001b[39m (scores \u001b[38;5;241m-\u001b[39m scores\u001b[38;5;241m.\u001b[39mmin()) \u001b[38;5;241m/\u001b[39m (scores\u001b[38;5;241m.\u001b[39mmax() \u001b[38;5;241m-\u001b[39m scores\u001b[38;5;241m.\u001b[39mmin())\n",
      "Input \u001b[0;32mIn [13]\u001b[0m, in \u001b[0;36mexport_to_UCSC_interact_hic.<locals>._value_to_color\u001b[0;34m(scores)\u001b[0m\n\u001b[1;32m    142\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_value_to_color\u001b[39m(scores):\n\u001b[1;32m    143\u001b[0m     S \u001b[38;5;241m=\u001b[39m (scores \u001b[38;5;241m-\u001b[39m scores\u001b[38;5;241m.\u001b[39mmin()) \u001b[38;5;241m/\u001b[39m (scores\u001b[38;5;241m.\u001b[39mmax() \u001b[38;5;241m-\u001b[39m scores\u001b[38;5;241m.\u001b[39mmin())\n\u001b[0;32m--> 144\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin([\u001b[38;5;28mstr\u001b[39m(x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m mapper\u001b[38;5;241m.\u001b[39mto_rgba(s, \u001b[38;5;28mbytes\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)][\u001b[38;5;241m0\u001b[39m:\u001b[38;5;241m3\u001b[39m]) \u001b[38;5;28;01mfor\u001b[39;00m s \u001b[38;5;129;01min\u001b[39;00m S]\n",
      "Input \u001b[0;32mIn [13]\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    142\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_value_to_color\u001b[39m(scores):\n\u001b[1;32m    143\u001b[0m     S \u001b[38;5;241m=\u001b[39m (scores \u001b[38;5;241m-\u001b[39m scores\u001b[38;5;241m.\u001b[39mmin()) \u001b[38;5;241m/\u001b[39m (scores\u001b[38;5;241m.\u001b[39mmax() \u001b[38;5;241m-\u001b[39m scores\u001b[38;5;241m.\u001b[39mmin())\n\u001b[0;32m--> 144\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin([\u001b[38;5;28mstr\u001b[39m(x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m \u001b[43mmapper\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_rgba\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mbytes\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m][\u001b[38;5;241m0\u001b[39m:\u001b[38;5;241m3\u001b[39m]) \u001b[38;5;28;01mfor\u001b[39;00m s \u001b[38;5;129;01min\u001b[39;00m S]\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/matplotlib/cm.py:455\u001b[0m, in \u001b[0;36mScalarMappable.to_rgba\u001b[0;34m(self, x, alpha, bytes, norm)\u001b[0m\n\u001b[1;32m    453\u001b[0m x \u001b[38;5;241m=\u001b[39m ma\u001b[38;5;241m.\u001b[39masarray(x)\n\u001b[1;32m    454\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m norm:\n\u001b[0;32m--> 455\u001b[0m     x \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnorm\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    456\u001b[0m rgba \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcmap(x, alpha\u001b[38;5;241m=\u001b[39malpha, \u001b[38;5;28mbytes\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mbytes\u001b[39m)\n\u001b[1;32m    457\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m rgba\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/matplotlib/colors.py:1254\u001b[0m, in \u001b[0;36mNormalize.__call__\u001b[0;34m(self, value, clip)\u001b[0m\n\u001b[1;32m   1252\u001b[0m     resdat \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39mdata\n\u001b[1;32m   1253\u001b[0m     resdat \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m=\u001b[39m vmin\n\u001b[0;32m-> 1254\u001b[0m     resdat \u001b[38;5;241m/\u001b[39m\u001b[38;5;241m=\u001b[39m (vmax \u001b[38;5;241m-\u001b[39m vmin)\n\u001b[1;32m   1255\u001b[0m     result \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mma\u001b[38;5;241m.\u001b[39marray(resdat, mask\u001b[38;5;241m=\u001b[39mresult\u001b[38;5;241m.\u001b[39mmask, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   1256\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_scalar:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import pandas as pd\n",
    "cell_lines=['Scenicplus-importance_links_all', 'Scenicplus-rho_links_all']\n",
    "color_dict={'Scenicplus-importance_links_all':'Blues', 'Scenicplus-rho_links_all':'Reds'}\n",
    "save_path = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files_final/'\n",
    "species = 'hsapiens'\n",
    "assembly = 'hg38'\n",
    "path_bedToBigBed = '/data/leuven/software/biomed/haswell_centos7/2018a/software/Kent_tools/20190730-linux.x86_64/bin/'\n",
    "biomart_host = 'http://oct2016.archive.ensembl.org/'\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/formatted/'\n",
    "for cell_line in cell_lines:\n",
    "    files = [f for f in listdir(path) if isfile(join(path, f))]\n",
    "    file = [f for f in files if cell_line in f]\n",
    "    hic_data = pd.read_csv(path+file[0], sep='\\t')\n",
    "    hic_data.columns = ['Chromosome', 'Start', 'End', 'target', 'importance', 'TSS', 'Strand', 'Distance', 'region']\n",
    "    hic_data['rho'] = hic_data['importance']\n",
    "    hic_data = hic_data[['target', 'region', 'importance', 'rho', 'Distance']]\n",
    "    hic_data = hic_data[hic_data['region'].isin(scplus_obj.region_names)]\n",
    "    scplus_obj.uns['region_to_gene'] = hic_data\n",
    "    r2g_data = export_to_UCSC_interact_hic(scplus_obj,\n",
    "                    species,\n",
    "                    os.path.join(save_path,cell_line+'.links.all.bed'),\n",
    "                    path_bedToBigBed=path_bedToBigBed,\n",
    "                    bigbed_outfile=os.path.join(save_path,cell_line+'.links.all.bb'),\n",
    "                    region_to_gene_key='region_to_gene',\n",
    "                    pbm_host=biomart_host,\n",
    "                    assembly=assembly,\n",
    "                    ucsc_track_name='R2G',\n",
    "                    ucsc_description='SCENIC+ region to gene links',\n",
    "                    cmap_neg='Reds',\n",
    "                    cmap_pos=color_dict[cell_line],\n",
    "                    key_for_color='importance',\n",
    "                    scale_by_gene=True,\n",
    "                    subset_for_eRegulons_regions=False,\n",
    "                    eRegulons_key='eRegulons')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0324021d-2a8b-4e73-a594-2d718cb7754e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-01-28 15:50:47,701 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-01-28 15:50:48,195 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/venv/lib/python3.8/site-packages/pybiomart/dataset.py:269: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  result = pd.read_csv(StringIO(response.text), sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-01-28 15:56:13,377 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files/HepG2.hic.all.bed\n",
      "2022-01-28 15:56:27,256 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files/HepG2.hic.all.notscaled.bb\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "#from scenicplus.enhancer_to_gene import export_to_UCSC_interact \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "62e415d0-4daa-4bdb-9ccd-59c8beaaaa4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>importance</th>\n",
       "      <th>rho</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21625214-21625715</td>\n",
       "      <td>13.353646</td>\n",
       "      <td>13.353646</td>\n",
       "      <td>228095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21482852-21483353</td>\n",
       "      <td>13.769301</td>\n",
       "      <td>13.769301</td>\n",
       "      <td>85733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21483548-21484049</td>\n",
       "      <td>13.769301</td>\n",
       "      <td>13.769301</td>\n",
       "      <td>86429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21507880-21508381</td>\n",
       "      <td>14.197148</td>\n",
       "      <td>14.197148</td>\n",
       "      <td>110761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ZNF493</td>\n",
       "      <td>chr19:21474600-21475101</td>\n",
       "      <td>19.167118</td>\n",
       "      <td>19.167118</td>\n",
       "      <td>77481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576819</th>\n",
       "      <td>PRR23D2</td>\n",
       "      <td>chr8:27616977-27617478</td>\n",
       "      <td>2559.157700</td>\n",
       "      <td>2559.157700</td>\n",
       "      <td>-19835564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576825</th>\n",
       "      <td>FAM90A8P</td>\n",
       "      <td>chr8:30469420-30469921</td>\n",
       "      <td>2584.061000</td>\n",
       "      <td>2584.061000</td>\n",
       "      <td>22731033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576839</th>\n",
       "      <td>GRAPL</td>\n",
       "      <td>chr17:19021510-19022011</td>\n",
       "      <td>3444.935300</td>\n",
       "      <td>3444.935300</td>\n",
       "      <td>-106025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576844</th>\n",
       "      <td>SLX1A</td>\n",
       "      <td>chr16:22170680-22171181</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>-8023195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576846</th>\n",
       "      <td>BOLA2B</td>\n",
       "      <td>chr16:22170680-22171181</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>3885.789000</td>\n",
       "      <td>8023626</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>651842 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           target                   region   importance          rho  Distance\n",
       "10         ZNF493  chr19:21625214-21625715    13.353646    13.353646    228095\n",
       "12         ZNF493  chr19:21482852-21483353    13.769301    13.769301     85733\n",
       "13         ZNF493  chr19:21483548-21484049    13.769301    13.769301     86429\n",
       "15         ZNF493  chr19:21507880-21508381    14.197148    14.197148    110761\n",
       "24         ZNF493  chr19:21474600-21475101    19.167118    19.167118     77481\n",
       "...           ...                      ...          ...          ...       ...\n",
       "2576819   PRR23D2   chr8:27616977-27617478  2559.157700  2559.157700 -19835564\n",
       "2576825  FAM90A8P   chr8:30469420-30469921  2584.061000  2584.061000  22731033\n",
       "2576839     GRAPL  chr17:19021510-19022011  3444.935300  3444.935300   -106025\n",
       "2576844     SLX1A  chr16:22170680-22171181  3885.789000  3885.789000  -8023195\n",
       "2576846    BOLA2B  chr16:22170680-22171181  3885.789000  3885.789000   8023626\n",
       "\n",
       "[651842 rows x 5 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hic_data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0210917c-c8d3-47c9-9139-13bb7fb77f18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-01-28 15:48:02,833 R2G          INFO     Downloading gene annotation from biomart, using dataset: hsapiens_gene_ensembl\n",
      "2022-01-28 15:48:03,532 R2G          INFO     Formatting data ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/venv/lib/python3.8/site-packages/pybiomart/dataset.py:269: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  result = pd.read_csv(StringIO(response.text), sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-01-28 15:49:45,835 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files/HepG2.hic.eGRN.bed\n",
      "2022-01-28 15:49:49,890 R2G          INFO     Writing data to: /staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/HiC/bb_files/HepG2.hic.eGRN.notscaled.bb\n"
     ]
    }
   ],
   "source": [
    "eRegulon_regions = list(set(flatten_list([ereg.target_regions for ereg in scplus_obj.uns['eRegulons']])))\n",
    "hic_data = hic_data[hic_data['region'].isin(eRegulon_regions)]\n",
    "scplus_obj.uns['region_to_gene'] = hic_data\n",
    "r2g_data = export_to_UCSC_interact_hic(scplus_obj,\n",
    "                    species,\n",
    "                    os.path.join(save_path,'HepG2.hic.eGRN.bed'),\n",
    "                    path_bedToBigBed=path_bedToBigBed,\n",
    "                    bigbed_outfile=os.path.join(save_path,'HepG2.hic.eGRN.notscaled.bb'),\n",
    "                    region_to_gene_key='region_to_gene',\n",
    "                    pbm_host=biomart_host,\n",
    "                    assembly=assembly,\n",
    "                    ucsc_track_name='R2G',\n",
    "                    ucsc_description='SCENIC+ region to gene links',\n",
    "                    cmap_neg='Reds',\n",
    "                    cmap_pos='Greens',\n",
    "                    key_for_color='importance',\n",
    "                    scale_by_gene=False,\n",
    "                    subset_for_eRegulons_regions=False,\n",
    "                    eRegulons_key='eRegulons')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f0b9117-2026-45bd-a324-ea3cdd718d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the downstream analyses\n",
    "\n",
    "#outDir = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final/'\n",
    "outDir = '/data/users/cbravo/DPCL/scenicplus_final_autoreg/'\n",
    "import pickle\n",
    "infile = open(outDir+'scplus_obj.pkl', 'rb')\n",
    "scplus_obj = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "import pickle\n",
    "infile = open('/data/users/cbravo/DPCL/scenicplus_final_autoreg/genie3/region_ranking.pkl', 'rb')\n",
    "region_ranking = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "import pickle\n",
    "infile = open('/data/users/cbravo/DPCL/scenicplus_final_autoreg/genie3/gene_ranking.pkl', 'rb')\n",
    "gene_ranking = pickle.load(infile)\n",
    "infile.close()\n",
    "\n",
    "from scenicplus.wrappers.run_scenicplus import *\n",
    "run_scenicplus_genie3(scplus_obj,\n",
    "    variable = ['ACC_Cell_type'],\n",
    "    species = 'hsapiens',\n",
    "    assembly = 'hg38',\n",
    "    tf_file = '/data/users/cbravo/resources/allTFs_hg38.txt',\n",
    "    save_path = '/data/users/cbravo/DPCL/scenicplus_final_autoreg/genie3/',\n",
    "    biomart_host = 'http://oct2016.archive.ensembl.org/',\n",
    "    upstream = [1000, 150000],\n",
    "    downstream = [1000, 150000],   \n",
    "    calculate_TF_eGRN_correlation = False,\n",
    "    calculate_DEGs_DARs = True,\n",
    "    export_to_loom_file = True,\n",
    "    export_to_UCSC_file = True,\n",
    "    region_ranking=region_ranking,\n",
    "    gene_ranking=gene_ranking,\n",
    "    tree_structure = ('DPCL', 'SCENIC+', 'genie3'),\n",
    "    path_bedToBigBed = '/media/data/users/cbravo/software/KENT/',\n",
    "    n_cpu = 20,\n",
    "    _temp_dir = '/media/data/users/cbravo/ray_spill'\n",
    "    )\n",
    "\n",
    "from scenicplus.wrappers.run_scenicplus import *\n",
    "run_scenicplus_genie3(scplus_obj,\n",
    "    variable = ['ACC_Cell_type'],\n",
    "    species = 'hsapiens',\n",
    "    assembly = 'hg38',\n",
    "    tf_file = '/staging/leuven/stg_00002/lcb/cflerin/resources/allTFs_hg38.txt',\n",
    "    save_path = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/scenicplus_final_autoreg/genie3/',\n",
    "    biomart_host = 'http://oct2016.archive.ensembl.org/',\n",
    "    upstream = [1000, 150000],\n",
    "    downstream = [1000, 150000],   \n",
    "    calculate_TF_eGRN_correlation = False,\n",
    "    calculate_DEGs_DARs = True,\n",
    "    export_to_loom_file = True,\n",
    "    export_to_UCSC_file = True,\n",
    "    tree_structure = ('DPCL', 'SCENIC+', 'genie3'),\n",
    "    path_bedToBigBed = '/data/leuven/software/biomed/haswell_centos7/2018a/software/Kent_tools/20190730-linux.x86_64/bin/',\n",
    "    n_cpu = 14,\n",
    "    _temp_dir = '/scratch/leuven/313/vsc31305/ray_spill'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "159e3271-7212-4c73-ae16-6a58f1f4c0e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.scenicplus_class import SCENICPLUS, create_SCENICPLUS_object\n",
    "from scenicplus.preprocessing.filtering import *\n",
    "from scenicplus.cistromes import *\n",
    "from scenicplus.enhancer_to_gene import get_search_space, calculate_regions_to_genes_relationships, RF_KWARGS\n",
    "from scenicplus.enhancer_to_gene import export_to_UCSC_interact \n",
    "from scenicplus.utils import format_egrns, export_eRegulons\n",
    "from scenicplus.eregulon_enrichment import *\n",
    "from scenicplus.TF_to_gene import *\n",
    "from scenicplus.grn_builder.gsea_approach import build_grn\n",
    "from scenicplus.dimensionality_reduction import *\n",
    "from scenicplus.RSS import *\n",
    "from scenicplus.diff_features import *\n",
    "from scenicplus.loom import *\n",
    "from typing import Dict, List, Mapping, Optional, Sequence\n",
    "import os\n",
    "import dill\n",
    "import time\n",
    "\n",
    "def run_scenicplus_genie3(scplus_obj: 'SCENICPLUS',\n",
    "    variable: List[str],\n",
    "    species: str,\n",
    "    assembly: str,\n",
    "    tf_file: str,\n",
    "    save_path: str,\n",
    "    biomart_host: Optional[str] = 'http://www.ensembl.org',\n",
    "    upstream: Optional[List] = [1000, 150000],\n",
    "    downstream: Optional[List] = [1000, 150000],\n",
    "    region_ranking: Optional['CisTopicImputedFeatures'] = None,\n",
    "    gene_ranking: Optional['CisTopicImputedFeatures'] = None,\n",
    "    calculate_TF_eGRN_correlation: Optional[bool] = True,\n",
    "    calculate_DEGs_DARs: Optional[bool] = True,\n",
    "    export_to_loom_file: Optional[bool] = True,\n",
    "    export_to_UCSC_file: Optional[bool] = True,\n",
    "    tree_structure: Sequence[str] = (),\n",
    "    path_bedToBigBed: Optional[str] = None,\n",
    "    n_cpu: Optional[int] = 1,\n",
    "    _temp_dir: Optional[str] = '/scratch/leuven/313/vsc31305/ray_spill'\n",
    "    ):\n",
    "    \"\"\"\n",
    "    Wrapper to run SCENIC+\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    scplus_obj: `class::SCENICPLUS`\n",
    "        A SCENICPLUS object.\n",
    "    variables: List[str]\n",
    "        Variables to use for RSS, TF-eGRN correlation and markers.\n",
    "    species: str\n",
    "        Species from which data comes from. Possible values: 'hsapiens', 'mmusculus', 'dmelanogaster'\n",
    "    assembly: str\n",
    "        Genome assembly to which the data was mapped. Possible values: 'hg38'\n",
    "    tf_file: str\n",
    "        Path to file containing genes that are TFs\n",
    "    save_path: str\n",
    "        Folder in which results will be saved\n",
    "    biomart_host: str, optional\n",
    "        Path to biomart host. Make sure that the host matches your genome assembly\n",
    "    upstream: str, optional\n",
    "        Upstream space to use for region to gene relationships\n",
    "    downstream: str, optional\n",
    "        Upstream space to use for region to gene relationships\n",
    "    region_ranking: `class::CisTopicImputedFeatures`, optional\n",
    "        Precomputed region ranking\n",
    "    gene_ranking: `class::CisTopicImputedFeatures`, optional\n",
    "        Precomputed gene ranking\n",
    "    calculate_TF_eGRN_correlation: bool, optional\n",
    "        Whether to calculate the TF-eGRN correlation based on the variables\n",
    "    calculate_DEGs_DARs: bool, optional\n",
    "        Whether to calculate DARs/DEGs based on the variables\n",
    "    export_to_loom_file: bool, optional\n",
    "        Whether to export data to loom files (gene based/region based)\n",
    "    export_to_UCSC_file: bool, optional\n",
    "        Whether to export region-to-gene links and eregulons to bed files\n",
    "    tree_structure: sequence, optional\n",
    "        Tree structure for loom files\n",
    "    path_bedToBigBed: str, optional\n",
    "        Path to convert bed files to big bed when exporting to UCSC (required if files are meant to be\n",
    "        used in a hub)\n",
    "    n_cpu: int, optional\n",
    "        Number of cores to use\n",
    "    _temp_dir: str, optional\n",
    "        Temporary directory for ray\n",
    "    \"\"\"\n",
    "    \n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = '%(asctime)s %(name)-12s %(levelname)-8s %(message)s'\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger('SCENIC+_wrapper')\n",
    "    \n",
    "    start_time = time.time()\n",
    "    \n",
    "    check_folder = os.path.isdir(save_path)\n",
    "    if not check_folder:\n",
    "        os.makedirs(save_path)\n",
    "        log.info(\"Created folder : \"+ save_path)\n",
    "\n",
    "    else:\n",
    "        log.info(save_path + \" folder already exists.\")\n",
    "    \n",
    "    if 'Cistromes' not in scplus_obj.uns.keys():\n",
    "        log.info('Merging cistromes')\n",
    "        merge_cistromes(scplus_obj)\n",
    "    \n",
    "    \n",
    "    if 'search_space' not in scplus_obj.uns.keys():\n",
    "        log.info('Getting search space')\n",
    "        get_search_space(scplus_obj,\n",
    "                     biomart_host = biomart_host,\n",
    "                     species = species,\n",
    "                     assembly = assembly, \n",
    "                     upstream = upstream,\n",
    "                     downstream = downstream)\n",
    "                 \n",
    "    if 'region_to_gene' not in scplus_obj.uns.keys():\n",
    "        log.info('Inferring region to gene relationships')\n",
    "        calculate_regions_to_genes_relationships(scplus_obj, \n",
    "                        ray_n_cpu = n_cpu, \n",
    "                        _temp_dir = _temp_dir,\n",
    "                        importance_scoring_method = 'RF',\n",
    "                        importance_scoring_kwargs = RF_KWARGS)\n",
    "                        \n",
    "    if 'TF2G_adj' not in scplus_obj.uns.keys():\n",
    "        log.info('Inferring TF to gene relationships')\n",
    "        calculate_TFs_to_genes_relationships(scplus_obj, \n",
    "                        tf_file = tf_file,\n",
    "                        ray_n_cpu = n_cpu, \n",
    "                        method = 'GBM',\n",
    "                        _temp_dir = _temp_dir,\n",
    "                        key= 'TF2G_adj')\n",
    "                        \n",
    "    if 'eRegulons' not in scplus_obj.uns.keys():\n",
    "        log.info('Build eGRN')\n",
    "        build_grn(scplus_obj,\n",
    "                 min_target_genes = 10,\n",
    "                 adj_pval_thr = 1,\n",
    "                 min_regions_per_gene = 0,\n",
    "                 quantiles = (0.85, 0.90, 0.95),\n",
    "                 top_n_regionTogenes_per_gene = (5, 10, 15),\n",
    "                 top_n_regionTogenes_per_region = (),\n",
    "                 binarize_using_basc = True,\n",
    "                 rho_dichotomize_tf2g = True,\n",
    "                 rho_dichotomize_r2g = True,\n",
    "                 rho_dichotomize_eregulon = True,\n",
    "                 rho_threshold = 0.05,\n",
    "                 keep_extended_motif_annot = True,\n",
    "                 merge_eRegulons = True, \n",
    "                 order_regions_to_genes_by = 'importance',\n",
    "                 order_TFs_to_genes_by = 'importance',\n",
    "                 key_added = 'eRegulons',\n",
    "                 cistromes_key = 'Unfiltered',\n",
    "                 disable_tqdm = True, \n",
    "                 ray_n_cpu = n_cpu,\n",
    "                 _temp_dir = _temp_dir)\n",
    "\n",
    "    if 'eRegulon_metadata' not in scplus_obj.uns.keys():\n",
    "        log.info('Formatting eGRNs')\n",
    "        format_egrns(scplus_obj,\n",
    "                      eregulons_key = 'eRegulons',\n",
    "                      TF2G_key = 'TF2G_adj',\n",
    "                      key_added = 'eRegulon_metadata')\n",
    "\n",
    "\n",
    "    if 'eRegulon_signatures' not in scplus_obj.uns.keys():\n",
    "        log.info('Converting eGRNs to signatures')\n",
    "        get_eRegulons_as_signatures(scplus_obj,\n",
    "                                     eRegulon_metadata_key='eRegulon_metadata', \n",
    "                                     key_added='eRegulon_signatures')\n",
    "\n",
    "#if 'eRegulon_AUC' not in scplus_obj.uns.keys():\n",
    "    log.info('Calculating eGRNs AUC')\n",
    "    if region_ranking is None:\n",
    "        log.info('Calculating region ranking')\n",
    "        region_ranking = make_rankings(scplus_obj, target='region')\n",
    "        with open(os.path.join(save_path,'region_ranking.pkl'), 'wb') as f:\n",
    "            dill.dump(region_ranking, f)\n",
    "    log.info('Calculating eGRNs region based AUC')\n",
    "    score_eRegulons(scplus_obj,\n",
    "            ranking = region_ranking,\n",
    "            eRegulon_signatures_key = 'eRegulon_signatures',\n",
    "            key_added = 'eRegulon_AUC', \n",
    "            enrichment_type= 'region',\n",
    "            auc_threshold = 0.05,\n",
    "            normalize = False,\n",
    "            n_cpu = n_cpu)\n",
    "    if gene_ranking is None:\n",
    "        log.info('Calculating gene ranking')\n",
    "        gene_ranking = make_rankings(scplus_obj, target='gene')\n",
    "        with open(os.path.join(save_path,'gene_ranking.pkl'), 'wb') as f:\n",
    "            dill.dump(gene_ranking, f)\n",
    "    log.info('Calculating eGRNs gene based AUC')\n",
    "    score_eRegulons(scplus_obj,\n",
    "            gene_ranking,\n",
    "            eRegulon_signatures_key = 'eRegulon_signatures',\n",
    "            key_added = 'eRegulon_AUC', \n",
    "            enrichment_type = 'gene',\n",
    "            auc_threshold = 0.05,\n",
    "            normalize= False,\n",
    "            n_cpu = n_cpu)\n",
    "                \n",
    "    if calculate_TF_eGRN_correlation is True:\n",
    "        log.info('Calculating TF-eGRNs AUC correlation')\n",
    "        for var in variable:\n",
    "            generate_pseudobulks(scplus_obj, \n",
    "                             variable = var,\n",
    "                             auc_key = 'eRegulon_AUC',\n",
    "                             signature_key = 'Gene_based',\n",
    "                             nr_cells = 5,\n",
    "                             nr_pseudobulks = 100,\n",
    "                             seed=555)\n",
    "            generate_pseudobulks(scplus_obj, \n",
    "                                     variable = var,\n",
    "                                     auc_key = 'eRegulon_AUC',\n",
    "                                     signature_key = 'Region_based',\n",
    "                                     nr_cells = 5,\n",
    "                                     nr_pseudobulks = 100,\n",
    "                                     seed=555)\n",
    "            TF_cistrome_correlation(scplus_obj,\n",
    "                            variable = var, \n",
    "                            auc_key = 'eRegulon_AUC',\n",
    "                            signature_key = 'Gene_based',\n",
    "                            out_key = var+'_eGRN_gene_based')\n",
    "            TF_cistrome_correlation(scplus_obj,\n",
    "                                    variable = var, \n",
    "                                    auc_key = 'eRegulon_AUC',\n",
    "                                    signature_key = 'Region_based',\n",
    "                                    out_key = var+'_eGRN_region_based')\n",
    "                                \n",
    "#if 'eRegulon_AUC_thresholds' not in scplus_obj.uns.keys():\n",
    "    log.info('Binarizing eGRNs AUC')\n",
    "    binarize_AUC(scplus_obj, \n",
    "         auc_key='eRegulon_AUC',\n",
    "         out_key='eRegulon_AUC_thresholds',\n",
    "         signature_keys=['Gene_based', 'Region_based'],\n",
    "         n_cpu=n_cpu)\n",
    "\n",
    "#if 'eRegulons_UMAP' not in scplus_obj.dr_cell.keys():\n",
    "    log.info('Making eGRNs AUC UMAP')\n",
    "    run_eRegulons_umap(scplus_obj,\n",
    "               scale=True, signature_keys=['Gene_based', 'Region_based'])\n",
    "#if 'eRegulons_tSNE' not in scplus_obj.dr_cell.keys():\n",
    "    log.info('Making eGRNs AUC tSNE')\n",
    "    run_eRegulons_tsne(scplus_obj,\n",
    "               scale=True, signature_keys=['Gene_based', 'Region_based'])\n",
    "\n",
    "#if 'RSS' not in scplus_obj.uns.keys():\n",
    "    log.info('Calculating eRSS')\n",
    "    for var in variable:\n",
    "        regulon_specificity_scores(scplus_obj, \n",
    "                     var,\n",
    "                     signature_keys=['Gene_based'],\n",
    "                     out_key_suffix='_gene_based',\n",
    "                     scale=False)\n",
    "        regulon_specificity_scores(scplus_obj, \n",
    "                     var,\n",
    "                     signature_keys=['Region_based'],\n",
    "                     out_key_suffix='_region_based',\n",
    "                         scale=False)\n",
    "                         \n",
    "    if calculate_DEGs_DARs is True:\n",
    "        log.info('Calculating DEGs/DARs')\n",
    "        for var in variable:\n",
    "            get_differential_features(scplus_obj, var, use_hvg = True, contrast_type = ['DEGs', 'DARs'])\n",
    "            \n",
    "    if export_to_loom_file is True:\n",
    "        log.info('Exporting to loom file')\n",
    "        export_to_loom(scplus_obj, \n",
    "               signature_key = 'Gene_based',\n",
    "               tree_structure = tree_structure,\n",
    "               title =  'Gene based eGRN',\n",
    "               nomenclature = assembly,\n",
    "               out_fname=os.path.join(save_path,'SCENIC+_gene_based.loom'))\n",
    "        export_to_loom(scplus_obj, \n",
    "               signature_key = 'Region_based',\n",
    "               tree_structure = tree_structure,\n",
    "               title =  'Region based eGRN',\n",
    "               nomenclature = assembly,\n",
    "               out_fname=os.path.join(save_path,'SCENIC+_region_based.loom'))\n",
    "               \n",
    "    if export_to_UCSC_file is True:\n",
    "        log.info('Exporting to UCSC')\n",
    "        r2g_data = export_to_UCSC_interact(scplus_obj,\n",
    "                            species,\n",
    "                            os.path.join(save_path,'r2g.rho.bed'),\n",
    "                            path_bedToBigBed=path_bedToBigBed,\n",
    "                            bigbed_outfile=os.path.join(save_path,'r2g.rho.bb'),\n",
    "                            region_to_gene_key='region_to_gene',\n",
    "                            pbm_host=biomart_host,\n",
    "                            assembly=assembly,\n",
    "                            ucsc_track_name='R2G',\n",
    "                            ucsc_description='SCENIC+ region to gene links',\n",
    "                            cmap_neg='Reds',\n",
    "                            cmap_pos='Greens',\n",
    "                            key_for_color='rho',\n",
    "                            scale_by_gene=False,\n",
    "                            subset_for_eRegulons_regions=True,\n",
    "                            eRegulons_key='eRegulons')\n",
    "        r2g_data = export_to_UCSC_interact(scplus_obj,\n",
    "                            species,\n",
    "                            os.path.join(save_path,'r2g.importance.bed'),\n",
    "                            path_bedToBigBed=path_bedToBigBed,\n",
    "                            bigbed_outfile=os.path.join(save_path,'r2g.importance.bb'),\n",
    "                            region_to_gene_key='region_to_gene',\n",
    "                            pbm_host=biomart_host,\n",
    "                            assembly=assembly,\n",
    "                            ucsc_track_name='R2G',\n",
    "                            ucsc_description='SCENIC+ region to gene links',\n",
    "                            cmap_neg='Reds',\n",
    "                            cmap_pos='Greens',\n",
    "                            key_for_color='importance',\n",
    "                            scale_by_gene=True,\n",
    "                            subset_for_eRegulons_regions=True,\n",
    "                            eRegulons_key='eRegulons')\n",
    "        regions = export_eRegulons(scplus_obj,\n",
    "                os.path.join(save_path,'eRegulons.bed'),\n",
    "                assembly,\n",
    "                bigbed_outfile = os.path.join(save_path,'eRegulons.bb'),\n",
    "                eRegulon_metadata_key = 'eRegulon_metadata',\n",
    "                eRegulon_signature_key = 'eRegulon_signatures',\n",
    "                path_bedToBigBed=path_bedToBigBed)\n",
    "        \n",
    "    log.info('Saving object')         \n",
    "    with open(os.path.join(save_path,'scplus_obj_genie3.pkl'), 'wb') as f:\n",
    "        dill.dump(scplus_obj, f)\n",
    "\n",
    "    log.info('Finished! Took {} minutes'.format((time.time() - start_time)/60))       \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b95ac6db-1ec0-4b14-bf1c-0e8dbe445aef",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DEV IMAGE - SCENIC+ latest (use at own risk)",
   "language": "python",
   "name": "pyscenic_plus_dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
