{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "597d0c39-81d2-49df-9ad9-ac008e94e02e",
   "metadata": {},
   "source": [
    "### V10 - Clustered - cistarget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "553d9061-54a1-495a-b1b4-c078cbab2b6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# r06i01n13\n",
    "singularity exec scenicplus.sif ipython3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e9fe58d4-ade8-42fa-a0c5-194ca2197002",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "infile = open('/data/users/cbravo/motif_enrichment_benchmark/region_sets.pkl', 'rb')\n",
    "region_sets = pickle.load(infile)\n",
    "infile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e65d5f-323a-4048-bb0a-bf255ca55cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cisTarget - All\n",
    "# Load cistarget functions\n",
    "from pycistarget.motif_enrichment_cistarget import *\n",
    "for i in range(0,100):\n",
    "    print(i)\n",
    "    key = list(region_sets.keys())[i]\n",
    "    # Preload db, you can also just provide the path to the db. Preloading the database is useful if you want to test different parameters.\n",
    "    db = '/data/users/cbravo/hg38_SCREEN_cluster_db_public/cluster_SCREEN_public.regions_vs_motifs.rankings.feather'  \n",
    "    rg = {key: region_sets[key]}\n",
    "    ctx_db = cisTargetDatabase(db, rg)    \n",
    "    # Run (r06i00n12)\n",
    "    cistarget_dict = run_cistarget(ctx_db = ctx_db,\n",
    "                                   region_sets = rg,\n",
    "                                   specie = 'homo_sapiens',\n",
    "                                   auc_threshold = 0.005,\n",
    "                                   nes_threshold = 3.0,\n",
    "                                   rank_threshold = 0.05,\n",
    "                                   path_to_motif_annotations = '/data/users/cbravo/motif_enrichment_benchmark/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv',\n",
    "                                   n_cpu = 1,\n",
    "                                   _temp_dir='/data/users/cbravo/ray_spill',\n",
    "                                   annotation_version = 'v10nr_clust')\n",
    "    # Save pkl\n",
    "    import pickle\n",
    "    with open('/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/output_dicts/' + key + '_cisTarget_dict.pkl', 'wb') as f:\n",
    "      pickle.dump(cistarget_dict, f)\n",
    "    # Save html\n",
    "    out_file = '/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/html/'+ key + '.html'\n",
    "    cistarget_dict[key].motif_enrichment.to_html(open(out_file, 'w'), escape=False, col_space=80)\n",
    "    \n",
    "for i in range(100,200):\n",
    "    print(i)\n",
    "    key = list(region_sets.keys())[i]\n",
    "    # Preload db, you can also just provide the path to the db. Preloading the database is useful if you want to test different parameters.\n",
    "    db = '/data/users/cbravo/hg38_SCREEN_cluster_db_public/cluster_SCREEN_public.regions_vs_motifs.rankings.feather'  \n",
    "    rg = {key: region_sets[key]}\n",
    "    ctx_db = cisTargetDatabase(db, rg)    \n",
    "    # Run (r06i00n12)\n",
    "    cistarget_dict = run_cistarget(ctx_db = ctx_db,\n",
    "                                   region_sets = rg,\n",
    "                                   specie = 'homo_sapiens',\n",
    "                                   auc_threshold = 0.005,\n",
    "                                   nes_threshold = 3.0,\n",
    "                                   rank_threshold = 0.05,\n",
    "                                   path_to_motif_annotations = '/data/users/cbravo/motif_enrichment_benchmark/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv',\n",
    "                                   n_cpu = 1,\n",
    "                                   _temp_dir='/data/users/cbravo/ray_spill',\n",
    "                                   annotation_version = 'v10nr_clust')\n",
    "    # Save pkl\n",
    "    import pickle\n",
    "    with open('/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/output_dicts/' + key + '_cisTarget_dict.pkl', 'wb') as f:\n",
    "      pickle.dump(cistarget_dict, f)\n",
    "    # Save html\n",
    "    out_file = '/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/html/'+ key + '.html'\n",
    "    cistarget_dict[key].motif_enrichment.to_html(open(out_file, 'w'), escape=False, col_space=80)\n",
    "    \n",
    "for i in range(200,len(region_sets)):\n",
    "    print(i)\n",
    "    key = list(region_sets.keys())[i]\n",
    "    # Preload db, you can also just provide the path to the db. Preloading the database is useful if you want to test different parameters.\n",
    "    db = '/data/users/cbravo/hg38_SCREEN_cluster_db_public/cluster_SCREEN_public.regions_vs_motifs.rankings.feather'  \n",
    "    rg = {key: region_sets[key]}\n",
    "    ctx_db = cisTargetDatabase(db, rg)    \n",
    "    # Run (r06i00n12)\n",
    "    cistarget_dict = run_cistarget(ctx_db = ctx_db,\n",
    "                                   region_sets = rg,\n",
    "                                   specie = 'homo_sapiens',\n",
    "                                   auc_threshold = 0.005,\n",
    "                                   nes_threshold = 3.0,\n",
    "                                   rank_threshold = 0.05,\n",
    "                                   path_to_motif_annotations = '/data/users/cbravo/motif_enrichment_benchmark/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv',\n",
    "                                   n_cpu = 1,\n",
    "                                   _temp_dir='/data/users/cbravo/ray_spill',\n",
    "                                   annotation_version = 'v10nr_clust')\n",
    "    # Save pkl\n",
    "    import pickle\n",
    "    with open('/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/output_dicts/' + key + '_cisTarget_dict.pkl', 'wb') as f:\n",
    "      pickle.dump(cistarget_dict, f)\n",
    "    # Save html\n",
    "    out_file = '/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/html/'+ key + '.html'\n",
    "    cistarget_dict[key].motif_enrichment.to_html(open(out_file, 'w'), escape=False, col_space=80)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f709d8e-92c5-48cc-85fe-f810e4ee0be6",
   "metadata": {},
   "source": [
    "### V10 - Clustered - DEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42c30cef-a003-4c27-a59e-999740219b2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# r06i01n13\n",
    "singularity exec scenicplus.sif ipython3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1bf00f7a-3155-4c60-ba32-643fee5a1c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "infile = open('/data/users/cbravo/motif_enrichment_benchmark/region_sets.pkl', 'rb')\n",
    "region_sets = pickle.load(infile)\n",
    "infile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fe97c76-1807-45e7-a0c1-03da171d9671",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare annotation\n",
    "import pybiomart as pbm\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "biomart_host = 'http://www.ensembl.org'\n",
    "promoter_space = 500\n",
    "name = 'hsapiens_gene_ensembl'\n",
    "dem_log2fc_thr = 0.5,\n",
    "dem_motif_hit_thr = 3.0,\n",
    "dem_max_bg_regions = 500,\n",
    "dataset = pbm.Dataset(name=name,  host=biomart_host)\n",
    "annot = dataset.query(attributes=['chromosome_name', 'transcription_start_site', 'strand', 'external_gene_name', 'transcript_biotype'])\n",
    "annot.columns = ['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "annot['Chromosome'] = annot['Chromosome'].astype('str')\n",
    "filterf = annot['Chromosome'].str.contains('CHR|GL|JH|MT')\n",
    "annot = annot[~filterf]\n",
    "annot['Chromosome'] = annot['Chromosome'].str.replace(r'(\\b\\S)', r'chr\\1')\n",
    "annot = annot[annot.Transcript_type == 'protein_coding']\n",
    "annot = annot.dropna(subset = ['Chromosome', 'Start'])\n",
    "# Check if chromosomes have chr\n",
    "check = region_sets[list(region_sets.keys())[0]]\n",
    "if not any(['chr' in c for c in check[list(check.keys())[0]].df['Chromosome']]):\n",
    "    annot.Chromosome = annot.Chromosome.str.replace('chr', '')\n",
    "annot_dem=annot.copy()\n",
    "# Define promoter space\n",
    "annot['End'] = annot['Start'].astype(int)+promoter_space\n",
    "annot['Start'] = annot['Start'].astype(int)-promoter_space\n",
    "annot = pr.PyRanges(annot[['Chromosome', 'Start', 'End']])\n",
    "region_set_list = [region_sets[x].df for x in region_sets.keys()]\n",
    "all_regions = pr.PyRanges(pd.concat(region_set_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aef00af-1b03-49b1-af48-14e631343e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEM\n",
    "# Load DEM functions\n",
    "from pycistarget.motif_enrichment_dem import *\n",
    "for i in range(0, 100):\n",
    "    print(i)\n",
    "    key = list(region_sets.keys())[i]\n",
    "    # Create bg\n",
    "    bg_overlap = all_regions.count_overlaps(region_sets[key])\n",
    "    bg = bg_overlap[bg_overlap.NumberOverlaps == 0]\n",
    "    bg = pr.PyRanges(bg.df.sample(frac=0.0005))\n",
    "    # Preload db, you can also just provide the path to the db. Preloading the database is useful if you want to test different parameters.\n",
    "    db = '/data/users/cbravo/hg38_SCREEN_cluster_db_public/cluster_SCREEN_public.regions_vs_motifs.scores.feather'  \n",
    "    rg = {key: region_sets[key], 'Background': bg}\n",
    "    dem_db = DEMDatabase(db, rg)    \n",
    "    # Run (r06i00n12)\n",
    "    DEM_dict = DEM(dem_db = dem_db,\n",
    "                                   region_sets = rg,\n",
    "                                   contrasts = [[[key],['Background']]],\n",
    "                                   log2fc_thr = 0.5,\n",
    "                                   motif_hit_thr = 3.0,\n",
    "                                   specie = 'homo_sapiens',\n",
    "                                   max_bg_regions = 5000,\n",
    "                                   genome_annotation = annot_dem,\n",
    "                                   promoter_space = promoter_space,\n",
    "                                   path_to_motif_annotations = '/data/users/cbravo/motif_enrichment_benchmark/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv',\n",
    "                                   n_cpu = 1,\n",
    "                                   tmp_dir = '/data/users/cbravo/ray_spill',\n",
    "                                   _temp_dir= '/data/users/cbravo/ray_spill',\n",
    "                                   annotation_version = 'v10nr_clust')\n",
    "    # Save pkl\n",
    "    import pickle\n",
    "    with open('/data/users/cbravo/motif_enrichment_benchmark/DEM_v10_public/output_dicts/' + key + '_DEM_dict.pkl', 'wb') as f:\n",
    "      pickle.dump(DEM_dict, f)\n",
    "    # Save html\n",
    "    out_file = '/data/users/cbravo/motif_enrichment_benchmark/DEM_v10_public/html/'+ key + '.html'\n",
    "    DEM_dict.motif_enrichment[key+'_VS_Background'].to_html(open(out_file, 'w'), escape=False, col_space=80)\n",
    "    \n",
    "# DEM\n",
    "# Load DEM functions\n",
    "from pycistarget.motif_enrichment_dem import *\n",
    "for i in range(100,200):\n",
    "    print(i)\n",
    "    key = list(region_sets.keys())[i]\n",
    "    # Create bg\n",
    "    bg_overlap = all_regions.count_overlaps(region_sets[key])\n",
    "    bg = bg_overlap[bg_overlap.NumberOverlaps == 0]\n",
    "    bg = pr.PyRanges(bg.df.sample(frac=0.0005))\n",
    "    # Preload db, you can also just provide the path to the db. Preloading the database is useful if you want to test different parameters.\n",
    "    db = '/data/users/cbravo/hg38_SCREEN_cluster_db_public/cluster_SCREEN_public.regions_vs_motifs.scores.feather'  \n",
    "    rg = {key: region_sets[key], 'Background': bg}\n",
    "    dem_db = DEMDatabase(db, rg)    \n",
    "    # Run (r06i00n12)\n",
    "    DEM_dict = DEM(dem_db = dem_db,\n",
    "                                   region_sets = rg,\n",
    "                                   contrasts = [[[key],['Background']]],\n",
    "                                   log2fc_thr = 0.5,\n",
    "                                   motif_hit_thr = 3.0,\n",
    "                                   specie = 'homo_sapiens',\n",
    "                                   max_bg_regions = 5000,\n",
    "                                   genome_annotation = annot_dem,\n",
    "                                   promoter_space = promoter_space,\n",
    "                                   path_to_motif_annotations = '/data/users/cbravo/motif_enrichment_benchmark/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv',\n",
    "                                   n_cpu = 1,\n",
    "                                   tmp_dir = '/data/users/cbravo/ray_spill',\n",
    "                                   _temp_dir= '/data/users/cbravo/ray_spill',\n",
    "                                   annotation_version = 'v10nr_clust')\n",
    "    # Save pkl\n",
    "    import pickle\n",
    "    with open('/data/users/cbravo/motif_enrichment_benchmark/DEM_v10_public/output_dicts/' + key + '_DEM_dict.pkl', 'wb') as f:\n",
    "      pickle.dump(DEM_dict, f)\n",
    "    # Save html\n",
    "    out_file = '/data/users/cbravo/motif_enrichment_benchmark/DEM_v10_public/html/'+ key + '.html'\n",
    "    DEM_dict.motif_enrichment[key+'_VS_Background'].to_html(open(out_file, 'w'), escape=False, col_space=80)\n",
    "# DEM\n",
    "# Load DEM functions\n",
    "from pycistarget.motif_enrichment_dem import *\n",
    "for i in range(200,len(region_sets)):\n",
    "    print(i)\n",
    "    key = list(region_sets.keys())[i]\n",
    "    # Create bg\n",
    "    bg_overlap = all_regions.count_overlaps(region_sets[key])\n",
    "    bg = bg_overlap[bg_overlap.NumberOverlaps == 0]\n",
    "    bg = pr.PyRanges(bg.df.sample(frac=0.0005))\n",
    "    # Preload db, you can also just provide the path to the db. Preloading the database is useful if you want to test different parameters.\n",
    "    db = '/data/users/cbravo/hg38_SCREEN_cluster_db_public/cluster_SCREEN_public.regions_vs_motifs.scores.feather'  \n",
    "    rg = {key: region_sets[key], 'Background': bg}\n",
    "    dem_db = DEMDatabase(db, rg)    \n",
    "    # Run (r06i00n12)\n",
    "    DEM_dict = DEM(dem_db = dem_db,\n",
    "                                   region_sets = rg,\n",
    "                                   contrasts = [[[key],['Background']]],\n",
    "                                   log2fc_thr = 0.5,\n",
    "                                   motif_hit_thr = 3.0,\n",
    "                                   specie = 'homo_sapiens',\n",
    "                                   max_bg_regions = 5000,\n",
    "                                   genome_annotation = annot_dem,\n",
    "                                   promoter_space = promoter_space,\n",
    "                                   path_to_motif_annotations = '/data/users/cbravo/motif_enrichment_benchmark/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv',\n",
    "                                   n_cpu = 1,\n",
    "                                   tmp_dir = '/data/users/cbravo/ray_spill',\n",
    "                                   _temp_dir= '/data/users/cbravo/ray_spill',\n",
    "                                   annotation_version = 'v10nr_clust')\n",
    "    # Save pkl\n",
    "    import pickle\n",
    "    with open('/data/users/cbravo/motif_enrichment_benchmark/DEM_v10_public/output_dicts/' + key + '_DEM_dict.pkl', 'wb') as f:\n",
    "      pickle.dump(DEM_dict, f)\n",
    "    # Save html\n",
    "    out_file = '/data/users/cbravo/motif_enrichment_benchmark/DEM_v10_public/html/'+ key + '.html'\n",
    "    DEM_dict.motif_enrichment[key+'_VS_Background'].to_html(open(out_file, 'w'), escape=False, col_space=80)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cdec597-8a8b-4998-ab8a-ee7bd6ef3fad",
   "metadata": {},
   "source": [
    "# Prepare data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a2aca3-99e0-4153-ade3-ded44cae821e",
   "metadata": {},
   "source": [
    "## cisTarget clustered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8941825b-89d3-44a0-a5cd-087eb62d498b",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/'\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "dicts = listdir(path+'output_dicts')\n",
    "dire = path+'tsv/'\n",
    "\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    menr[list(menr.keys())[0]].motif_enrichment.to_csv(dire+list(menr.keys())[0]+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bedbd8eb-d0eb-45ce-aa2f-f14dcb55cbd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path='/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/'\n",
    "dire = '/data/users/cbravo/motif_enrichment_benchmark/cisTarget_v10_public/cistromes_hits/'\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    for key in menr.keys():\n",
    "        print(key)\n",
    "        tf = key.split('_')[1]\n",
    "        dict_keys = menr[key].cistromes['Region_set'].keys()\n",
    "        name = [name for name in dict_keys if tf+'_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr[key].cistromes['Region_set'][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)\n",
    "        name = [name for name in dict_keys if tf+'_extended_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr[key].cistromes['Region_set'][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c27a47ca-a647-49ad-b206-5e012e1bd653",
   "metadata": {},
   "source": [
    "## cisTarget unclustered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c1ea80ea-18d4-44b5-8560-15919ad795d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/cisTarget_v10_unclustered/'\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "dicts = listdir(path+'output_dicts')\n",
    "dire = path+'tsv/'\n",
    "\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    menr[list(menr.keys())[0]].motif_enrichment.to_csv(dire+list(menr.keys())[0]+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d893884-7fd8-49e2-83dc-8aac1d1e2eab",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/cisTarget_v10_unclustered/'\n",
    "dire = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/cisTarget_v10_unclustered/cistromes_hits/'\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    for key in menr.keys():\n",
    "        print(key)\n",
    "        tf = key.split('_')[1]\n",
    "        dict_keys = menr[key].cistromes['Region_set'].keys()\n",
    "        name = [name for name in dict_keys if tf+'_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr[key].cistromes['Region_set'][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)\n",
    "        name = [name for name in dict_keys if tf+'_extended_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr[key].cistromes['Region_set'][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93b620df-56e7-4906-8975-0f858ab88f3e",
   "metadata": {},
   "source": [
    "## cisTarget archetypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7fb307c5-aef2-44cc-b32c-7c1678b3ec6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/cisTarget_v10_archetype/'\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "dicts = listdir(path+'output_dicts')\n",
    "dire = path+'tsv/'\n",
    "\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    menr[list(menr.keys())[0]].motif_enrichment.to_csv(dire+list(menr.keys())[0]+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cf7a3ac-db0a-464f-a40a-31ddc24ff2b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/cisTarget_v10_archetype/'\n",
    "dire = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/cisTarget_v10_archetype/cistromes_hits/'\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    for key in menr.keys():\n",
    "        print(key)\n",
    "        tf = key.split('_')[1]\n",
    "        dict_keys = menr[key].cistromes['Region_set'].keys()\n",
    "        name = [name for name in dict_keys if tf+'_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr[key].cistromes['Region_set'][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)\n",
    "        name = [name for name in dict_keys if tf+'_extended_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr[key].cistromes['Region_set'][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45ae82e2-e319-46ed-8090-9f8295e4f8c2",
   "metadata": {},
   "source": [
    "## DEM unclustered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "acf2adbc-f810-4868-8094-83a5c7e91af8",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_unclustered/'\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "dicts = listdir(path+'output_dicts')\n",
    "dire = path+'tsv/'\n",
    "\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    menr.motif_enrichment[list(menr.motif_enrichment.keys())[0]].to_csv(dire+list(menr.motif_enrichment.keys())[0]+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a76d85f5-940d-4249-8ffc-c9a684123cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_unclustered/'\n",
    "dire = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_unclustered/cistromes_hits/'\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    for key in menr.cistromes['Region_set'].keys():\n",
    "        print(key)\n",
    "        tf = key.split('_')[1]\n",
    "        dict_keys = menr.cistromes['Region_set'][key].keys()\n",
    "        name = [name for name in dict_keys if tf+'_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr.cistromes['Region_set'][key][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)\n",
    "        name = [name for name in dict_keys if tf+'_extended_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr.cistromes['Region_set'][key][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "117ddb03-2586-4890-a9ed-8826a10b00ad",
   "metadata": {},
   "source": [
    "## DEM archetype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d5cf43c4-a91d-497e-add0-98cbec7a16f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_archetype/'\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "dicts = listdir(path+'output_dicts')\n",
    "dire = path+'tsv/'\n",
    "\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    menr.motif_enrichment[list(menr.motif_enrichment.keys())[0]].to_csv(dire+list(menr.motif_enrichment.keys())[0]+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5baf264a-61fc-4a2b-8af0-d7271f55b948",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_archetype/'\n",
    "dire = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_archetype/cistromes_hits/'\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    for key in menr.cistromes['Region_set'].keys():\n",
    "        print(key)\n",
    "        tf = key.split('_')[1]\n",
    "        dict_keys = menr.cistromes['Region_set'][key].keys()\n",
    "        name = [name for name in dict_keys if tf+'_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr.cistromes['Region_set'][key][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)\n",
    "        name = [name for name in dict_keys if tf+'_extended_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr.cistromes['Region_set'][key][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3451a409-45d3-404c-8dba-54faa695bd7f",
   "metadata": {},
   "source": [
    "## DEM clustered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9c4b3730-4f63-4b85-8da6-c72473aa76b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_clustered/'\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "dicts = listdir(path+'output_dicts')\n",
    "dire = path+'tsv/'\n",
    "\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    menr.motif_enrichment[list(menr.motif_enrichment.keys())[0]].to_csv(dire+list(menr.motif_enrichment.keys())[0]+'.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cca5ec4c-574c-494a-be00-023a8cd55ade",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path='/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_clustered/'\n",
    "dire = '/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/ChIP-seq/motif_enrichment_benchmark/DEM_v10_clustered/cistromes_hits/'\n",
    "for i in range(len(dicts)):\n",
    "    print(i)\n",
    "    d = dicts[i]\n",
    "    f = path+'output_dicts/'+d\n",
    "    import pickle\n",
    "    infile = open(f, 'rb')\n",
    "    menr = pickle.load(infile)\n",
    "    infile.close()\n",
    "    for key in menr.cistromes['Region_set'].keys():\n",
    "        print(key)\n",
    "        tf = key.split('_')[1]\n",
    "        dict_keys = menr.cistromes['Region_set'][key].keys()\n",
    "        name = [name for name in dict_keys if tf+'_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr.cistromes['Region_set'][key][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)\n",
    "        name = [name for name in dict_keys if tf+'_extended_(' in name]\n",
    "        if len(name) == 1:\n",
    "            save_name = name[0].split('_(')[0]\n",
    "            pd.DataFrame(menr.cistromes['Region_set'][key][name[0]]).to_csv(dire+key+'_'+save_name+'.tsv', sep='\\t', header=False, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DEV IMAGE - SCENIC+ latest (use at own risk)",
   "language": "python",
   "name": "pyscenic_plus_dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
